// @iota/notarization/node/notarization_wasm.js
/* @ts-self-types="./notarization_wasm.d.ts" */
const { WasmIotaTransactionBlockResponseWrapper } = require(`@iota/iota-interaction-ts/node/iota_client_helpers`);
const { Ed25519PublicKey } = require(`@iota/iota-sdk/keypairs/ed25519`);
const { Secp256k1PublicKey } = require(`@iota/iota-sdk/keypairs/secp256k1`);
const { Secp256r1PublicKey } = require(`@iota/iota-sdk/keypairs/secp256r1`);
const { TransactionDataBuilder } = require(`@iota/iota-sdk/transactions`);
const { inspect } = require(`util`);

/**
 * Represents a transaction for creating dynamic notarization's.
 *
 * Dynamic notarization's can be updated after creation, with modification capabilities
 * for both state and metadata.
 */
class CreateNotarizationDynamic {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(CreateNotarizationDynamic.prototype);
        obj.__wbg_ptr = ptr;
        CreateNotarizationDynamicFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    toJSON() {
        return {
        };
    }
    toString() {
        return JSON.stringify(this);
    }
    [inspect.custom]() {
        return Object.assign(Object.create({constructor: this.constructor}), this.toJSON());
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        CreateNotarizationDynamicFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_createnotarizationdynamic_free(ptr, 0);
    }
    /**
     * Applies transaction effects and events to this notarization creation operation.
     *
     *
     * This method is called automatically by Transaction::build_programmable_transaction()
     * and Transaction::apply() methods after the transaction has been successfully submitted
     * to process the results from the ledger.
     *
     * # Arguments
     * * `effects` - The transaction block effects to apply.
     * * `events` - The transaction block events to apply.
     *
     * # Returns
     * The created notarization ID if successful.
     * @param {TransactionEffects} wasm_effects
     * @param {IotaEvent[]} wasm_events
     * @param {CoreClientReadOnly} client
     * @returns {Promise<OnChainNotarization>}
     */
    applyWithEvents(wasm_effects, wasm_events, client) {
        const ptr = this.__destroy_into_raw();
        const ret = wasm.createnotarizationdynamic_applyWithEvents(ptr, wasm_effects, wasm_events, client);
        return ret;
    }
    /**
     * Builds and returns a programmable transaction for creating a dynamic notarization.
     *
     * # Returns
     * The binary BCS serialization of the programmable transaction.
     * This transaction can be submitted to the network to create a new dynamic notarization.
     *
     * # Errors
     * Returns an error if the transaction cannot be built due to invalid parameters
     * or other constraints.
     * @param {CoreClientReadOnly} client
     * @returns {Promise<Uint8Array>}
     */
    buildProgrammableTransaction(client) {
        const ret = wasm.createnotarizationdynamic_buildProgrammableTransaction(this.__wbg_ptr, client);
        return ret;
    }
    /**
     * @param {NotarizationBuilderDynamic} builder
     */
    constructor(builder) {
        _assertClass(builder, NotarizationBuilderDynamic);
        var ptr0 = builder.__destroy_into_raw();
        const ret = wasm.createnotarizationdynamic_new(ptr0);
        this.__wbg_ptr = ret >>> 0;
        CreateNotarizationDynamicFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
}
if (Symbol.dispose) CreateNotarizationDynamic.prototype[Symbol.dispose] = CreateNotarizationDynamic.prototype.free;
exports.CreateNotarizationDynamic = CreateNotarizationDynamic;

/**
 * Represents a transaction for creating locked notarization's.
 *
 * Locked notarization's cannot be modified after creation, ensuring data permanence.
 */
class CreateNotarizationLocked {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(CreateNotarizationLocked.prototype);
        obj.__wbg_ptr = ptr;
        CreateNotarizationLockedFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    toJSON() {
        return {
        };
    }
    toString() {
        return JSON.stringify(this);
    }
    [inspect.custom]() {
        return Object.assign(Object.create({constructor: this.constructor}), this.toJSON());
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        CreateNotarizationLockedFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_createnotarizationlocked_free(ptr, 0);
    }
    /**
     * Applies transaction effects and events to this notarization creation operation.
     *
     * This method is called automatically by Transaction::build_programmable_transaction()
     * and Transaction::apply() methods after the transaction has been successfully submitted
     * to process the results from the ledger.
     *
     * # Arguments
     * * `effects` - The transaction block effects to apply.
     * * `events` - The transaction block events to apply.
     *
     * # Returns
     * The created notarization ID if successful.
     * @param {TransactionEffects} wasm_effects
     * @param {IotaEvent[]} wasm_events
     * @param {CoreClientReadOnly} client
     * @returns {Promise<OnChainNotarization>}
     */
    applyWithEvents(wasm_effects, wasm_events, client) {
        const ptr = this.__destroy_into_raw();
        const ret = wasm.createnotarizationlocked_applyWithEvents(ptr, wasm_effects, wasm_events, client);
        return ret;
    }
    /**
     * Builds and returns a programmable transaction for creating a locked notarization.
     *
     * # Returns
     * The binary BCS serialization of the programmable transaction.
     * This transaction can be submitted to the network to create a new locked notarization.
     *
     * # Errors
     * Returns an error if the transaction cannot be built due to invalid parameters
     * or other constraints.
     * @param {CoreClientReadOnly} client
     * @returns {Promise<Uint8Array>}
     */
    buildProgrammableTransaction(client) {
        const ret = wasm.createnotarizationlocked_buildProgrammableTransaction(this.__wbg_ptr, client);
        return ret;
    }
    /**
     * @param {NotarizationBuilderLocked} builder
     */
    constructor(builder) {
        _assertClass(builder, NotarizationBuilderLocked);
        var ptr0 = builder.__destroy_into_raw();
        const ret = wasm.createnotarizationlocked_new(ptr0);
        this.__wbg_ptr = ret >>> 0;
        CreateNotarizationLockedFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
}
if (Symbol.dispose) CreateNotarizationLocked.prototype[Symbol.dispose] = CreateNotarizationLocked.prototype.free;
exports.CreateNotarizationLocked = CreateNotarizationLocked;

/**
 * Represents the different types of data that can be notarized.
 */
class Data {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(Data.prototype);
        obj.__wbg_ptr = ptr;
        DataFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    toJSON() {
        return {
            value: this.value,
        };
    }
    toString() {
        return JSON.stringify(this);
    }
    [inspect.custom]() {
        return Object.assign(Object.create({constructor: this.constructor}), this.toJSON());
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        DataFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_data_free(ptr, 0);
    }
    /**
     * Converts the data to a byte array.
     *
     * # Returns
     * A `Uint8Array` containing the byte representation of the data.
     * @returns {Uint8Array}
     */
    toBytes() {
        const ret = wasm.data_toBytes(this.__wbg_ptr);
        var v1 = getArrayU8FromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        return v1;
    }
    /**
     * Converts the data to a string representation.
     *
     * # Returns
     * A `String` containing the text representation of the data.
     * @returns {string}
     */
    toString() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.data_toString(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Retrieves the value of the data as a `any`.
     *
     * # Returns
     * A `any` containing the data, either as bytes or text.
     * @returns {any}
     */
    get value() {
        const ret = wasm.data_value(this.__wbg_ptr);
        return ret;
    }
}
if (Symbol.dispose) Data.prototype[Symbol.dispose] = Data.prototype.free;
exports.Data = Data;

/**
 * A default implementation for {@link HttpClient}.
 */
class DefaultHttpClient {
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        DefaultHttpClientFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_defaulthttpclient_free(ptr, 0);
    }
    constructor() {
        const ret = wasm.defaulthttpclient_new();
        this.__wbg_ptr = ret >>> 0;
        DefaultHttpClientFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * @param {Request} request
     * @returns {Promise<Response>}
     */
    send(request) {
        const ret = wasm.defaulthttpclient_send(this.__wbg_ptr, request);
        return ret;
    }
}
if (Symbol.dispose) DefaultHttpClient.prototype[Symbol.dispose] = DefaultHttpClient.prototype.free;
exports.DefaultHttpClient = DefaultHttpClient;

/**
 * Represents a transaction for deleting a notarization.
 */
class DestroyNotarization {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(DestroyNotarization.prototype);
        obj.__wbg_ptr = ptr;
        DestroyNotarizationFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    toJSON() {
        return {
        };
    }
    toString() {
        return JSON.stringify(this);
    }
    [inspect.custom]() {
        return Object.assign(Object.create({constructor: this.constructor}), this.toJSON());
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        DestroyNotarizationFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_destroynotarization_free(ptr, 0);
    }
    /**
     * Applies transaction effects and events to this notarization delete operation.
     *
     * This method is called automatically by Transaction::build_programmable_transaction()
     * and Transaction::apply() methods after the transaction has been successfully submitted
     * to process the results from the ledger.
     *
     * # Arguments
     * * `effects` - The transaction block effects to apply.
     * * `events` - The transaction block events to apply.
     * @param {TransactionEffects} wasm_effects
     * @param {IotaEvent[]} wasm_events
     * @param {CoreClientReadOnly} client
     * @returns {Promise<Empty>}
     */
    applyWithEvents(wasm_effects, wasm_events, client) {
        const ptr = this.__destroy_into_raw();
        const ret = wasm.destroynotarization_applyWithEvents(ptr, wasm_effects, wasm_events, client);
        return ret;
    }
    /**
     * Builds and returns a programmable transaction for deleting a notarization.
     *
     * # Returns
     * The binary BCS serialization of the programmable transaction.
     * This transaction can be submitted to the network to delete a notarization.
     *
     * # Errors
     * Returns an error if the transaction cannot be built due to invalid parameters
     * or other constraints.
     * @param {CoreClientReadOnly} client
     * @returns {Promise<Uint8Array>}
     */
    buildProgrammableTransaction(client) {
        const ret = wasm.destroynotarization_buildProgrammableTransaction(this.__wbg_ptr, client);
        return ret;
    }
    /**
     * @param {string} object_id
     */
    constructor(object_id) {
        const ptr0 = passStringToWasm0(object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.destroynotarization_new(ptr0, len0);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        this.__wbg_ptr = ret[0] >>> 0;
        DestroyNotarizationFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
}
if (Symbol.dispose) DestroyNotarization.prototype[Symbol.dispose] = DestroyNotarization.prototype.free;
exports.DestroyNotarization = DestroyNotarization;

class Empty {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(Empty.prototype);
        obj.__wbg_ptr = ptr;
        EmptyFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    toJSON() {
        return {
        };
    }
    toString() {
        return JSON.stringify(this);
    }
    [inspect.custom]() {
        return Object.assign(Object.create({constructor: this.constructor}), this.toJSON());
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        EmptyFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_empty_free(ptr, 0);
    }
}
if (Symbol.dispose) Empty.prototype[Symbol.dispose] = Empty.prototype.free;
exports.Empty = Empty;

class GasStationParams {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(GasStationParams.prototype);
        obj.__wbg_ptr = ptr;
        GasStationParamsFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        GasStationParamsFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_gasstationparams_free(ptr, 0);
    }
    /**
     * @returns {bigint}
     */
    get gasReservationDuration() {
        const ret = wasm.gasstationparams_gasReservationDuration(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * @returns {HeaderMap}
     */
    get headers() {
        const ret = wasm.gasstationparams_headers(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return takeFromExternrefTable0(ret[0]);
    }
    /**
     * @param {GasStationParamsI | null} [params]
     */
    constructor(params) {
        const ret = wasm.gasstationparams_new(isLikeNone(params) ? 0 : addToExternrefTable0(params));
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        this.__wbg_ptr = ret[0] >>> 0;
        GasStationParamsFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * Adds an `Authorization` header using `token` as a bearer token.
     * @param {string} token
     * @returns {GasStationParams}
     */
    withAuthToken(token) {
        const ptr0 = passStringToWasm0(token, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.gasstationparams_withAuthToken(this.__wbg_ptr, ptr0, len0);
        return GasStationParams.__wrap(ret);
    }
}
if (Symbol.dispose) GasStationParams.prototype[Symbol.dispose] = GasStationParams.prototype.free;
exports.GasStationParams = GasStationParams;

/**
 * Represents immutable metadata of a notarization.
 */
class ImmutableMetadata {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(ImmutableMetadata.prototype);
        obj.__wbg_ptr = ptr;
        ImmutableMetadataFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    toJSON() {
        return {
            createdAt: this.createdAt,
            description: this.description,
            locking: this.locking,
        };
    }
    toString() {
        return JSON.stringify(this);
    }
    [inspect.custom]() {
        return Object.assign(Object.create({constructor: this.constructor}), this.toJSON());
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        ImmutableMetadataFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_immutablemetadata_free(ptr, 0);
    }
    /**
     * Retrieves the timestamp when the notarization was created.
     *
     * # Returns
     * The timestamp as `number` value representing the seconds since the Unix epoch.
     * @returns {bigint}
     */
    get createdAt() {
        const ret = wasm.immutablemetadata_createdAt(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * Retrieves the description of the notarization.
     *
     * # Returns
     * A description `string`, if existing.
     * @returns {string | undefined}
     */
    get description() {
        const ret = wasm.immutablemetadata_description(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Retrieves the optional lock metadata for the notarization.
     *
     * # Returns
     * A `LockMetadata` instance, if existing.
     * @returns {LockMetadata | undefined}
     */
    get locking() {
        const ret = wasm.immutablemetadata_locking(this.__wbg_ptr);
        return ret === 0 ? undefined : LockMetadata.__wrap(ret);
    }
}
if (Symbol.dispose) ImmutableMetadata.prototype[Symbol.dispose] = ImmutableMetadata.prototype.free;
exports.ImmutableMetadata = ImmutableMetadata;

/**
 * Represents the lock metadata of a notarization.
 */
class LockMetadata {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(LockMetadata.prototype);
        obj.__wbg_ptr = ptr;
        LockMetadataFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    toJSON() {
        return {
            deleteLock: this.deleteLock,
            transferLock: this.transferLock,
            updateLock: this.updateLock,
        };
    }
    toString() {
        return JSON.stringify(this);
    }
    [inspect.custom]() {
        return Object.assign(Object.create({constructor: this.constructor}), this.toJSON());
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        LockMetadataFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_lockmetadata_free(ptr, 0);
    }
    /**
     * The delete lock configuration.
     * @returns {TimeLock}
     */
    get deleteLock() {
        const ret = wasm.__wbg_get_lockmetadata_deleteLock(this.__wbg_ptr);
        return TimeLock.__wrap(ret);
    }
    /**
     * The transfer lock configuration.
     * @returns {TimeLock}
     */
    get transferLock() {
        const ret = wasm.__wbg_get_lockmetadata_transferLock(this.__wbg_ptr);
        return TimeLock.__wrap(ret);
    }
    /**
     * The update lock configuration.
     * @returns {TimeLock}
     */
    get updateLock() {
        const ret = wasm.__wbg_get_lockmetadata_updateLock(this.__wbg_ptr);
        return TimeLock.__wrap(ret);
    }
    /**
     * The delete lock configuration.
     * @param {TimeLock} arg0
     */
    set deleteLock(arg0) {
        _assertClass(arg0, TimeLock);
        var ptr0 = arg0.__destroy_into_raw();
        wasm.__wbg_set_lockmetadata_deleteLock(this.__wbg_ptr, ptr0);
    }
    /**
     * The transfer lock configuration.
     * @param {TimeLock} arg0
     */
    set transferLock(arg0) {
        _assertClass(arg0, TimeLock);
        var ptr0 = arg0.__destroy_into_raw();
        wasm.__wbg_set_lockmetadata_transferLock(this.__wbg_ptr, ptr0);
    }
    /**
     * The update lock configuration.
     * @param {TimeLock} arg0
     */
    set updateLock(arg0) {
        _assertClass(arg0, TimeLock);
        var ptr0 = arg0.__destroy_into_raw();
        wasm.__wbg_set_lockmetadata_updateLock(this.__wbg_ptr, ptr0);
    }
}
if (Symbol.dispose) LockMetadata.prototype[Symbol.dispose] = LockMetadata.prototype.free;
exports.LockMetadata = LockMetadata;

/**
 * Represents a builder for constructing dynamic notarization transactions.
 *
 * Dynamic notarizations are updatable records that can evolve over time.
 */
class NotarizationBuilderDynamic {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(NotarizationBuilderDynamic.prototype);
        obj.__wbg_ptr = ptr;
        NotarizationBuilderDynamicFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        NotarizationBuilderDynamicFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_notarizationbuilderdynamic_free(ptr, 0);
    }
    /**
     * Creates a new dynamic notarization builder.
     * @returns {NotarizationBuilderDynamic}
     */
    static dynamic() {
        const ret = wasm.notarizationbuilderdynamic_dynamic();
        return NotarizationBuilderDynamic.__wrap(ret);
    }
    /**
     * Finalizes the notarization builder and returns a transaction builder
     * that can be used to build and execute the final transaction on the ledger.
     *
     * # Returns
     * A `TransactionBuilder` to build and execute the transaction.
     * @returns {TransactionBuilder<CreateNotarizationDynamic>}
     */
    finish() {
        const ptr = this.__destroy_into_raw();
        const ret = wasm.notarizationbuilderdynamic_finish(ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return TransactionBuilder.__wrap(ret[0]);
    }
    /**
     * Adds a state to the notarization using binary data.
     *
     * # Arguments
     * * `data` - Binary data representing the state.
     * * `metadata` - Optional metadata associated with the state.
     * @param {Uint8Array} data
     * @param {string | null} [metadata]
     * @returns {NotarizationBuilderDynamic}
     */
    withBytesState(data, metadata) {
        const ptr = this.__destroy_into_raw();
        var ptr0 = isLikeNone(metadata) ? 0 : passStringToWasm0(metadata, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationbuilderdynamic_withBytesState(ptr, data, ptr0, len0);
        return NotarizationBuilderDynamic.__wrap(ret);
    }
    /**
     * Adds an immutable description to the notarization.
     *
     * # Arguments
     * * `description` - A string describing the notarization, or null to skip.
     * @param {string | null} [description]
     * @returns {NotarizationBuilderDynamic}
     */
    withImmutableDescription(description) {
        const ptr = this.__destroy_into_raw();
        var ptr0 = isLikeNone(description) ? 0 : passStringToWasm0(description, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationbuilderdynamic_withImmutableDescription(ptr, ptr0, len0);
        return NotarizationBuilderDynamic.__wrap(ret);
    }
    /**
     * Adds a state to the notarization using a string.
     *
     * # Arguments
     * * `data` - String data representing the state.
     * * `metadata` - Optional metadata associated with the state.
     * @param {string} data
     * @param {string | null} [metadata]
     * @returns {NotarizationBuilderDynamic}
     */
    withStringState(data, metadata) {
        const ptr = this.__destroy_into_raw();
        const ptr0 = passStringToWasm0(data, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        var ptr1 = isLikeNone(metadata) ? 0 : passStringToWasm0(metadata, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len1 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationbuilderdynamic_withStringState(ptr, ptr0, len0, ptr1, len1);
        return NotarizationBuilderDynamic.__wrap(ret);
    }
    /**
     * Adds a transfer lock to the notarization.
     *
     * # Arguments
     * * `lock` - A `TimeLock` specifying the transfer lock.
     * @param {TimeLock} lock
     * @returns {NotarizationBuilderDynamic}
     */
    withTransferLock(lock) {
        const ptr = this.__destroy_into_raw();
        _assertClass(lock, TimeLock);
        var ptr0 = lock.__destroy_into_raw();
        const ret = wasm.notarizationbuilderdynamic_withTransferLock(ptr, ptr0);
        return NotarizationBuilderDynamic.__wrap(ret);
    }
    /**
     * Adds updatable metadata to the notarization.
     *
     * # Arguments
     * * `metadata` - A string representing the metadata, or null to skip.
     * @param {string | null} [metadata]
     * @returns {NotarizationBuilderDynamic}
     */
    withUpdatableMetadata(metadata) {
        const ptr = this.__destroy_into_raw();
        var ptr0 = isLikeNone(metadata) ? 0 : passStringToWasm0(metadata, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationbuilderdynamic_withUpdatableMetadata(ptr, ptr0, len0);
        return NotarizationBuilderDynamic.__wrap(ret);
    }
}
if (Symbol.dispose) NotarizationBuilderDynamic.prototype[Symbol.dispose] = NotarizationBuilderDynamic.prototype.free;
exports.NotarizationBuilderDynamic = NotarizationBuilderDynamic;

/**
 * Represents a builder for constructing locked notarization transactions.
 *
 * Locked notarizations are immutable records (the notarization state) that cannot be modified
 * after creation.
 */
class NotarizationBuilderLocked {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(NotarizationBuilderLocked.prototype);
        obj.__wbg_ptr = ptr;
        NotarizationBuilderLockedFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    toJSON() {
        return {
        };
    }
    toString() {
        return JSON.stringify(this);
    }
    [inspect.custom]() {
        return Object.assign(Object.create({constructor: this.constructor}), this.toJSON());
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        NotarizationBuilderLockedFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_notarizationbuilderlocked_free(ptr, 0);
    }
    /**
     * Finalizes the notarization builder and returns a transaction builder
     * that can be used to build and execute the final transaction on the ledger.
     *
     * # Returns
     * A `TransactionBuilder` to build and execute the transaction.
     * @returns {TransactionBuilder<CreateNotarizationLocked>}
     */
    finish() {
        const ptr = this.__destroy_into_raw();
        const ret = wasm.notarizationbuilderlocked_finish(ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return TransactionBuilder.__wrap(ret[0]);
    }
    /**
     * Creates a new locked notarization builder.
     * @returns {NotarizationBuilderLocked}
     */
    static locked() {
        const ret = wasm.notarizationbuilderlocked_locked();
        return NotarizationBuilderLocked.__wrap(ret);
    }
    /**
     * Adds a state to the notarization using binary data.
     *
     * # Arguments
     * * `data` - Binary data representing the state.
     * * `metadata` - Optional metadata associated with the state.
     * @param {Uint8Array} data
     * @param {string | null} [metadata]
     * @returns {NotarizationBuilderLocked}
     */
    withBytesState(data, metadata) {
        const ptr = this.__destroy_into_raw();
        var ptr0 = isLikeNone(metadata) ? 0 : passStringToWasm0(metadata, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationbuilderlocked_withBytesState(ptr, data, ptr0, len0);
        return NotarizationBuilderLocked.__wrap(ret);
    }
    /**
     * Adds a delete lock to the notarization.
     *
     * # Arguments
     * * `lock` - A `TimeLock` specifying the delete lock.
     * @param {TimeLock} lock
     * @returns {NotarizationBuilderLocked}
     */
    withDeleteLock(lock) {
        const ptr = this.__destroy_into_raw();
        _assertClass(lock, TimeLock);
        var ptr0 = lock.__destroy_into_raw();
        const ret = wasm.notarizationbuilderlocked_withDeleteLock(ptr, ptr0);
        return NotarizationBuilderLocked.__wrap(ret);
    }
    /**
     * Adds an immutable description to the notarization.
     *
     * # Arguments
     * * `description` - A string describing the notarization, or null to skip.
     * @param {string | null} [description]
     * @returns {NotarizationBuilderLocked}
     */
    withImmutableDescription(description) {
        const ptr = this.__destroy_into_raw();
        var ptr0 = isLikeNone(description) ? 0 : passStringToWasm0(description, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationbuilderlocked_withImmutableDescription(ptr, ptr0, len0);
        return NotarizationBuilderLocked.__wrap(ret);
    }
    /**
     * Adds a state to the notarization using a string.
     *
     * # Arguments
     * * `data` - String data representing the state.
     * * `metadata` - Optional metadata associated with the state.
     * @param {string} data
     * @param {string | null} [metadata]
     * @returns {NotarizationBuilderLocked}
     */
    withStringState(data, metadata) {
        const ptr = this.__destroy_into_raw();
        const ptr0 = passStringToWasm0(data, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        var ptr1 = isLikeNone(metadata) ? 0 : passStringToWasm0(metadata, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len1 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationbuilderlocked_withStringState(ptr, ptr0, len0, ptr1, len1);
        return NotarizationBuilderLocked.__wrap(ret);
    }
    /**
     * Adds updatable metadata to the notarization.
     *
     * # Arguments
     * * `metadata` - A string representing the metadata, or null to skip.
     * @param {string | null} [metadata]
     * @returns {NotarizationBuilderLocked}
     */
    withUpdatableMetadata(metadata) {
        const ptr = this.__destroy_into_raw();
        var ptr0 = isLikeNone(metadata) ? 0 : passStringToWasm0(metadata, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationbuilderlocked_withUpdatableMetadata(ptr, ptr0, len0);
        return NotarizationBuilderLocked.__wrap(ret);
    }
}
if (Symbol.dispose) NotarizationBuilderLocked.prototype[Symbol.dispose] = NotarizationBuilderLocked.prototype.free;
exports.NotarizationBuilderLocked = NotarizationBuilderLocked;

/**
 * A client to interact with Notarization objects on the IOTA ledger.
 *
 * This client is used for read and write operations. For read-only capabilities,
 * you can use {@link NotarizationClientReadOnly}, which does not require an account or signing capabilities.
 */
class NotarizationClient {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(NotarizationClient.prototype);
        obj.__wbg_ptr = ptr;
        NotarizationClientFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        NotarizationClientFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_notarizationclient_free(ptr, 0);
    }
    /**
     * Creates a new notarization client.
     *
     * # Arguments
     * * `client` - A read-only notarization client.
     * * `signer` - A transaction signer for signing operations.
     *
     * # Returns
     * A `TransactionBuilder` to build and execute the transaction.
     * @param {NotarizationClientReadOnly} client
     * @param {TransactionSigner} signer
     * @returns {Promise<NotarizationClient>}
     */
    static create(client, signer) {
        _assertClass(client, NotarizationClientReadOnly);
        var ptr0 = client.__destroy_into_raw();
        const ret = wasm.notarizationclient_create(ptr0, signer);
        return ret;
    }
    /**
     * Creates a notarization builder which can be used to create a dynamic notarization.
     *
     * # Returns
     * A `NotarizationBuilderDynamic` instance.
     * @returns {NotarizationBuilderDynamic}
     */
    createDynamic() {
        const ret = wasm.notarizationclient_createDynamic(this.__wbg_ptr);
        return NotarizationBuilderDynamic.__wrap(ret);
    }
    /**
     * Creates a notarization builder which can be used to create a locked notarization.
     *
     * # Returns
     * A `NotarizationBuilderLocked` instance.
     * @returns {NotarizationBuilderLocked}
     */
    createLocked() {
        const ret = wasm.notarizationclient_createLocked(this.__wbg_ptr);
        return NotarizationBuilderLocked.__wrap(ret);
    }
    /**
     * Creates a transaction to destroy a notarization object on the ledger.
     *
     * # Arguments
     * * `notarization_id` - The ID of the notarization object to destroy.
     *
     * # Returns
     * A `TransactionBuilder` to build and execute the transaction.
     * @param {string} notarization_id
     * @returns {TransactionBuilder}
     */
    destroy(notarization_id) {
        const ptr0 = passStringToWasm0(notarization_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclient_destroy(this.__wbg_ptr, ptr0, len0);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return TransactionBuilder.__wrap(ret[0]);
    }
    /**
     * Retrieves the IOTA client instance.
     *
     * # Returns
     * The `IotaClient` instance.
     * @returns {IotaClient}
     */
    iotaClient() {
        const ret = wasm.notarizationclient_iotaClient(this.__wbg_ptr);
        return ret;
    }
    /**
     * Retrieves the network identifier.
     *
     * # Returns
     * The network identifier as a `string`.
     * @returns {string}
     */
    network() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.notarizationclient_network(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Retrieves the package history.
     *
     * # Returns
     * An `Array<string>` containing the package history.
     * @returns {string[]}
     */
    packageHistory() {
        const ret = wasm.notarizationclient_packageHistory(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Retrieves the package ID.
     *
     * # Returns
     * The package ID as a `string`.
     * @returns {string}
     */
    packageId() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.notarizationclient_packageId(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Retrieves a read-only version of the notarization client.
     *
     * # Returns
     * A `NotarizationClientReadOnly` instance.
     * @returns {NotarizationClientReadOnly}
     */
    readOnly() {
        const ret = wasm.notarizationclient_readOnly(this.__wbg_ptr);
        return NotarizationClientReadOnly.__wrap(ret);
    }
    /**
     * Retrieves the sender's address.
     *
     * # Returns
     * The sender's address as an `IotaAddress`.
     * @returns {string}
     */
    senderAddress() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.notarizationclient_senderAddress(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Retrieves the sender's public key.
     *
     * # Returns
     * The sender's public key as `PublicKey`.
     * @returns {PublicKey}
     */
    senderPublicKey() {
        const ret = wasm.notarizationclient_senderPublicKey(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return takeFromExternrefTable0(ret[0]);
    }
    /**
     * Retrieves the transaction signer.
     *
     * # Returns
     * The `TransactionSigner` instance.
     * @returns {TransactionSigner}
     */
    signer() {
        const ret = wasm.notarizationclient_signer(this.__wbg_ptr);
        return ret;
    }
    /**
     * Creates a transaction to transfer a notarization object to a new owner.
     *
     * # Arguments
     * * `notarization_id` - The ID of the notarization object to transfer.
     * * `recipient` - The recipient's IOTA address.
     *
     * # Returns
     * A `TransactionBuilder` to build and execute the transaction.
     * @param {string} notarization_id
     * @param {string} recipient
     * @returns {TransactionBuilder}
     */
    transferNotarization(notarization_id, recipient) {
        const ptr0 = passStringToWasm0(notarization_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ptr1 = passStringToWasm0(recipient, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len1 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclient_transferNotarization(this.__wbg_ptr, ptr0, len0, ptr1, len1);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return TransactionBuilder.__wrap(ret[0]);
    }
    /**
     * Creates a transaction to update the metadata of a notarization.
     *
     * **Important**: The `updatableMetadata` can only be updated depending on the used
     * `NotarizationMethod`:
     * - Dynamic: Can be updated anytime after notarization creation
     * - Locked: Immutable after notarization creation
     *
     * NOTE:
     * - does not affect the `stateVersionCount` or the `lastStateChangeAt` timestamp
     * - will fail if the notarization uses the `NotarizationMethod::Locked`
     * - Only the `updatableMetadata` can be changed; the `immutableMetadata::description` remains fixed
     * # Arguments
     * * `metadata` - The new metadata to update (optional).
     * * `notarization_id` - The ID of the notarization object.
     *
     * # Returns
     * A `TransactionBuilder` to build and execute the transaction.
     * @param {string | null | undefined} metadata
     * @param {string} notarization_id
     * @returns {TransactionBuilder}
     */
    updateMetadata(metadata, notarization_id) {
        var ptr0 = isLikeNone(metadata) ? 0 : passStringToWasm0(metadata, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        const ptr1 = passStringToWasm0(notarization_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len1 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclient_updateMetadata(this.__wbg_ptr, ptr0, len0, ptr1, len1);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return TransactionBuilder.__wrap(ret[0]);
    }
    /**
     * Creates a transaction to update the `state` of a notarization.
     *
     * **Important**: The `state` can only  be updated depending on the used `NotarizationMethod`:
     * - Dynamic: Can be updated anytime after notarization creation
     * - Locked: Immutable after notarization creation
     *
     * Using this function will:
     * - set the `state` to the `new_state`
     * - increase the `stateVersionCount` by 1
     * - set the `lastStateChangeAt` timestamp to the current clock  timestamp in milliseconds
     * - emits a `NotarizationUpdated` Move event in case of success
     * - fail if the notarization uses `NotarizationMethod` `Locked`
     *
     * # Arguments
     * * `new_state` - The new state to replace the current one.
     * * `notarization_id` - The ID of the notarization object.
     *
     * # Returns
     * A `TransactionBuilder` to build and execute the transaction.
     * @param {State} new_state
     * @param {string} notarization_id
     * @returns {TransactionBuilder}
     */
    updateState(new_state, notarization_id) {
        _assertClass(new_state, State);
        var ptr0 = new_state.__destroy_into_raw();
        const ptr1 = passStringToWasm0(notarization_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len1 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclient_updateState(this.__wbg_ptr, ptr0, ptr1, len1);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return TransactionBuilder.__wrap(ret[0]);
    }
}
if (Symbol.dispose) NotarizationClient.prototype[Symbol.dispose] = NotarizationClient.prototype.free;
exports.NotarizationClient = NotarizationClient;

/**
 * A client to interact with Notarization objects on the IOTA ledger.
 *
 * This client is used for read-only operations, meaning it does not require an account
 * or signing capabilities. For write operations, use {@link NotarizationClient}.
 */
class NotarizationClientReadOnly {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(NotarizationClientReadOnly.prototype);
        obj.__wbg_ptr = ptr;
        NotarizationClientReadOnlyFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        NotarizationClientReadOnlyFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_notarizationclientreadonly_free(ptr, 0);
    }
    /**
     * Retrieves the chain ID associated with this client.
     *
     * # Returns
     * A string representing the chain ID.
     * @returns {string}
     */
    chainId() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.notarizationclientreadonly_chainId(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Creates a new instance of `otarizationClientReadOnly`.
     *
     * # Arguments
     * * `iota_client` - The IOTA client used for interacting with the ledger.
     *
     * # Returns
     * A new `NotarizationClientReadOnly` instance.
     * @param {IotaClient} iota_client
     * @returns {Promise<NotarizationClientReadOnly>}
     */
    static create(iota_client) {
        const ret = wasm.notarizationclientreadonly_create(iota_client);
        return ret;
    }
    /**
     * Creates a new instance of `NotarizationClientReadOnly` using a specific package ID.
     *
     * # Arguments
     * * `iota_client` - The IOTA client used for interacting with the ledger.
     * * `iota_notarization_pkg_id` - The notarization package ID.
     *
     * # Returns
     * A new `NotarizationClientReadOnly` instance.
     * @param {IotaClient} iota_client
     * @param {string} iota_notarization_pkg_id
     * @returns {Promise<NotarizationClientReadOnly>}
     */
    static createWithPkgId(iota_client, iota_notarization_pkg_id) {
        const ptr0 = passStringToWasm0(iota_notarization_pkg_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclientreadonly_createWithPkgId(iota_client, ptr0, len0);
        return ret;
    }
    /**
     * Retrieves the creation timestamp for a notarization.
     *
     * # Arguments
     * * `notarized_object_id` - The ID of the notarization object.
     *
     * # Returns
     * The timestamp as `number` value representing the seconds since the Unix epoch.
     * @param {string} notarized_object_id
     * @returns {Promise<bigint>}
     */
    createdAtTs(notarized_object_id) {
        const ptr0 = passStringToWasm0(notarized_object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclientreadonly_createdAtTs(this.__wbg_ptr, ptr0, len0);
        return ret;
    }
    /**
     * Retrieves the description of a notarization.
     *
     * # Arguments
     * * `notarized_object_id` - The ID of a notarization object.
     *
     * # Returns
     * A description string, if existing.
     * @param {string} notarized_object_id
     * @returns {Promise<string | undefined>}
     */
    description(notarized_object_id) {
        const ptr0 = passStringToWasm0(notarized_object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclientreadonly_description(this.__wbg_ptr, ptr0, len0);
        return ret;
    }
    /**
     * Retrieves the [`OnChainNotarization`] of a notarized object.
     *
     * This method returns the on-chain notarization object for the given object ID.
     *
     * # Arguments
     * * `notarized_object_id` - The ID of the notarization object.
     *
     * # Returns
     * The [`OnChainNotarization`] object for the given object ID.
     * @param {string} notarized_object_id
     * @returns {Promise<OnChainNotarization>}
     */
    getNotarizationById(notarized_object_id) {
        const ptr0 = passStringToWasm0(notarized_object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclientreadonly_getNotarizationById(this.__wbg_ptr, ptr0, len0);
        return ret;
    }
    /**
     * Retrieves the underlying IOTA client used by this client.
     *
     * # Returns
     * The `IotaClient` instance.
     * @returns {IotaClient}
     */
    iotaClient() {
        const ret = wasm.notarizationclientreadonly_iotaClient(this.__wbg_ptr);
        return ret;
    }
    /**
     * Checks if destruction is allowed for a notarization object.
     *
     * # Arguments
     * * `notarized_object_id` - The ID of a notarization object.
     *
     * # Returns
     * A boolean indicating whether destruction is allowed.
     * False means that destroying is not allowed.
     * @param {string} notarized_object_id
     * @returns {Promise<boolean>}
     */
    isDestroyAllowed(notarized_object_id) {
        const ptr0 = passStringToWasm0(notarized_object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclientreadonly_isDestroyAllowed(this.__wbg_ptr, ptr0, len0);
        return ret;
    }
    /**
     * Checks if transferring a notarization object is allowed.
     *
     * # Arguments
     * * `notarized_object_id` - The ID of a notarization object.
     *
     * # Returns
     * A boolean indicating whether transfers are locked.
     * False means that transferring is allowed.
     * @param {string} notarized_object_id
     * @returns {Promise<boolean>}
     */
    isTransferLocked(notarized_object_id) {
        const ptr0 = passStringToWasm0(notarized_object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclientreadonly_isTransferLocked(this.__wbg_ptr, ptr0, len0);
        return ret;
    }
    /**
     * Checks if updates are locked for a notarization object.
     *
     * # Arguments
     * * `notarized_object_id` - The ID of a notarization object.
     *
     * # Returns
     * A boolean indicating whether updates are locked.
     * False means that updates are allowed.
     * @param {string} notarized_object_id
     * @returns {Promise<boolean>}
     */
    isUpdateLocked(notarized_object_id) {
        const ptr0 = passStringToWasm0(notarized_object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclientreadonly_isUpdateLocked(this.__wbg_ptr, ptr0, len0);
        return ret;
    }
    /**
     * Retrieves the timestamp of the last state change for a notarization.
     *
     * # Arguments
     * * `notarized_object_id` - The ID of the notarization object.
     *
     * # Returns
     * The timestamp as `number` value representing the seconds since the Unix epoch.
     * @param {string} notarized_object_id
     * @returns {Promise<bigint>}
     */
    lastStateChangeTs(notarized_object_id) {
        const ptr0 = passStringToWasm0(notarized_object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclientreadonly_lastStateChangeTs(this.__wbg_ptr, ptr0, len0);
        return ret;
    }
    /**
     * Retrieves the lock metadata of a notarization.
     *
     * # Arguments
     * * `notarized_object_id` - The ID of a notarization object.
     *
     * # Returns
     * The `LockMetadata`, if existing.
     * @param {string} notarized_object_id
     * @returns {Promise<LockMetadata | undefined>}
     */
    lockMetadata(notarized_object_id) {
        const ptr0 = passStringToWasm0(notarized_object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclientreadonly_lockMetadata(this.__wbg_ptr, ptr0, len0);
        return ret;
    }
    /**
     * Retrieves the network identifier associated with this client.
     *
     * # Returns
     * A string representing the network identifier.
     * @returns {string}
     */
    network() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.notarizationclientreadonly_network(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Retrieves the notarization method of a notarization.
     *
     * # Arguments
     * * `notarized_object_id` - The ID of a notarization object.
     *
     * # Returns
     * The `NotarizationMethod`.
     * @param {string} notarized_object_id
     * @returns {Promise<NotarizationMethod>}
     */
    notarizationMethod(notarized_object_id) {
        const ptr0 = passStringToWasm0(notarized_object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclientreadonly_notarizationMethod(this.__wbg_ptr, ptr0, len0);
        return ret;
    }
    /**
     * Retrieves the history of notarization package IDs.
     *
     * # Returns
     * An array of strings representing the package history.
     * @returns {string[]}
     */
    packageHistory() {
        const ret = wasm.notarizationclientreadonly_packageHistory(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Retrieves the package ID of the used notarization package.
     *
     * # Returns
     * A string representing the package ID.
     * @returns {string}
     */
    packageId() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.notarizationclientreadonly_packageId(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Retrieves the state of a notarization.
     *
     * # Arguments
     * * `notarized_object_id` - The ID of a notarization object.
     *
     * # Returns
     * The notarization `State`.
     * @param {string} notarized_object_id
     * @returns {Promise<State>}
     */
    state(notarized_object_id) {
        const ptr0 = passStringToWasm0(notarized_object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclientreadonly_state(this.__wbg_ptr, ptr0, len0);
        return ret;
    }
    /**
     * Retrieves the count of state versions for a notarization.
     *
     * # Arguments
     * * `notarized_object_id` - The ID of a notarization object.
     *
     * # Returns
     * Count as `number` value.
     * @param {string} notarized_object_id
     * @returns {Promise<bigint>}
     */
    stateVersionCount(notarized_object_id) {
        const ptr0 = passStringToWasm0(notarized_object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclientreadonly_stateVersionCount(this.__wbg_ptr, ptr0, len0);
        return ret;
    }
    /**
     * Retrieves the updatable metadata of a notarization.
     *
     * # Arguments
     * * `notarized_object_id` - The ID of a notarization object.
     *
     * # Returns
     * A metadata string, if existing.
     * @param {string} notarized_object_id
     * @returns {Promise<string | undefined>}
     */
    updatableMetadata(notarized_object_id) {
        const ptr0 = passStringToWasm0(notarized_object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.notarizationclientreadonly_updatableMetadata(this.__wbg_ptr, ptr0, len0);
        return ret;
    }
}
if (Symbol.dispose) NotarizationClientReadOnly.prototype[Symbol.dispose] = NotarizationClientReadOnly.prototype.free;
exports.NotarizationClientReadOnly = NotarizationClientReadOnly;

/**
 * A builder for creating IOTA Resource Locators (IRLs) pointing within a notarization.
 */
class NotarizationResourceBuilder {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(NotarizationResourceBuilder.prototype);
        obj.__wbg_ptr = ptr;
        NotarizationResourceBuilderFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        NotarizationResourceBuilderFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_notarizationresourcebuilder_free(ptr, 0);
    }
    /**
     * Returns an IRL referencing this {@link OnChainNotarization} state's data.
     * @returns {string}
     */
    data() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.notarizationresourcebuilder_data(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Returns an IRL referencing this {@link OnChainNotarization}'s immutable metadata.
     * @returns {string}
     */
    immutableMetadata() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.notarizationresourcebuilder_immutableMetadata(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Returns an IRL referencing this {@link OnChainNotarization}'s owner.
     * @returns {string}
     */
    owner() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.notarizationresourcebuilder_owner(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Returns an IRL referencing this {@link OnChainNotarization} state's metadata.
     * @returns {string}
     */
    stateMetadata() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.notarizationresourcebuilder_stateMetadata(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Returns an IRL referencing this {@link OnChainNotarization}'s updatable metadata.
     * @returns {string}
     */
    updatableMetadata() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.notarizationresourcebuilder_updatableMetadata(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
}
if (Symbol.dispose) NotarizationResourceBuilder.prototype[Symbol.dispose] = NotarizationResourceBuilder.prototype.free;
exports.NotarizationResourceBuilder = NotarizationResourceBuilder;

/**
 * Represents an on-chain notarization object.
 *
 * Provides access to various properties of the notarization, such as its ID, state, metadata
 * and method.
 */
class OnChainNotarization {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(OnChainNotarization.prototype);
        obj.__wbg_ptr = ptr;
        OnChainNotarizationFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    toJSON() {
        return {
            id: this.id,
            immutableMetadata: this.immutableMetadata,
            lastStateChangeAt: this.lastStateChangeAt,
            method: this.method,
            owner: this.owner,
            state: this.state,
            stateVersionCount: this.stateVersionCount,
            updatableMetadata: this.updatableMetadata,
        };
    }
    toString() {
        return JSON.stringify(this);
    }
    [inspect.custom]() {
        return Object.assign(Object.create({constructor: this.constructor}), this.toJSON());
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        OnChainNotarizationFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_onchainnotarization_free(ptr, 0);
    }
    /**
     * Retrieves the ID of the notarization.
     *
     * # Returns
     * A hexadecimal string representing the notarization ID.
     * @returns {string}
     */
    get id() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.onchainnotarization_id(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Retrieves the immutable metadata of the notarization.
     *
     * NOTE:
     * - provides immutable information, assertions and guaranties for third parties
     * - `immutableMetadata` are automatically created at creation time and cannot be updated thereafter
     *
     * # Returns
     * An `ImmutableMetadata` object containing the metadata.
     * @returns {ImmutableMetadata}
     */
    get immutableMetadata() {
        const ret = wasm.onchainnotarization_immutableMetadata(this.__wbg_ptr);
        return ImmutableMetadata.__wrap(ret);
    }
    /**
     * Returns an IOTA Resource Locator (IRL) to the data stored within
     * this notarization object.
     *
     * The returned IRL will be in the form:
     * `iota:<network alias or genesis digest>/<notarization ID>/state/data`.
     * # Errors
     * Throws an error if the given `network` string is not a valid IOTA
     * network identifier.
     * @param {string} network
     * @returns {NotarizationResourceBuilder}
     */
    iotaResourceLocatorBuilder(network) {
        const ptr0 = passStringToWasm0(network, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.onchainnotarization_iotaResourceLocatorBuilder(this.__wbg_ptr, ptr0, len0);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return NotarizationResourceBuilder.__wrap(ret[0]);
    }
    /**
     * Retrieves the timestamp of the last state change.
     *
     * # Returns
     * A `number` value representing the timestamp,
     * the time in milliseconds since UNIX epoch.
     * @returns {bigint}
     */
    get lastStateChangeAt() {
        const ret = wasm.onchainnotarization_lastStateChangeAt(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * Retrieves the notarization method.
     *
     * # Returns
     * A `NotarizationMethod` object representing the method.
     * @returns {NotarizationMethod}
     */
    get method() {
        const ret = wasm.onchainnotarization_method(this.__wbg_ptr);
        return __wbindgen_enum_NotarizationMethod[ret];
    }
    /**
     * Retrieves the owner address of the notarization.
     *
     * # Returns
     * An `IotaAddress` object representing the owner address.
     * @returns {string}
     */
    get owner() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.onchainnotarization_owner(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Retrieves the current `state` of the notarization.
     *
     * The `state` of a notarization contains the notarized data and metadata associated with
     * the current version of the `state`.
     *
     * `state` can be updated depending on the used `NotarizationMethod`:
     * - Dynamic: Can be updated anytime after notarization creation
     * - Locked: Immutable after notarization creation
     *
     * Use `NotarizationClient::updateState()` for `state` updates.
     *
     * # Returns
     * A `State` object representing the notarization state.
     * @returns {State}
     */
    get state() {
        const ret = wasm.onchainnotarization_state(this.__wbg_ptr);
        return State.__wrap(ret);
    }
    /**
     * Retrieves the count of state versions.
     *
     * # Returns
     * A `number` value representing the number of state versions.
     * @returns {bigint}
     */
    get stateVersionCount() {
        const ret = wasm.onchainnotarization_stateVersionCount(this.__wbg_ptr);
        return BigInt.asUintN(64, ret);
    }
    /**
     * Retrieves the updatable metadata of the notarization.
     *
     * Provides context or additional information for third parties
     *
     * `updatableMetadata` can be updated depending on the used `NotarizationMethod`:
     * - Dynamic: Can be updated anytime after notarization creation
     * - Locked: Immutable after notarization creation
     *
     * NOTE:
     * - `updatableMetadata` can be updated independently of `state`
     * - Updating `updatableMetadata` does not increase the `stateVersionCount`
     * - Updating `updatableMetadata` does not change the `lastStateChangeAt` timestamp
     * - Use `NotarizationClient::updateMetadata()` for `updatableMetadata` updates.
     *
     * # Returns
     * An optional string containing the metadata.
     * @returns {string | undefined}
     */
    get updatableMetadata() {
        const ret = wasm.onchainnotarization_updatableMetadata(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
}
if (Symbol.dispose) OnChainNotarization.prototype[Symbol.dispose] = OnChainNotarization.prototype.free;
exports.OnChainNotarization = OnChainNotarization;

/**
 * Represents the state of a notarization.
 *
 * State encapsulates the data being notarized along with optional metadata.
 * It serves as the primary content container for both locked and dynamic
 * notarizations.
 *
 * The notarization `State` can be updated by the owner depending on the used `NotarizationMethod`:
 * - Dynamic: `data` and `metadata` of the `State` can be updated anytime after creation
 * - Locked: The `State` is immutable after notarization creation
 *
 * `State` `data` and `metadata` can only be updated at once, using method `NotarizationClient::updateState()`
 * which will increase the `stateVersionCount` and update the `lastStateChangeAt`
 * timestamp of the notarization even if only the `metadata` are altered.
 */
class State {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(State.prototype);
        obj.__wbg_ptr = ptr;
        StateFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    toJSON() {
        return {
            data: this.data,
            metadata: this.metadata,
        };
    }
    toString() {
        return JSON.stringify(this);
    }
    [inspect.custom]() {
        return Object.assign(Object.create({constructor: this.constructor}), this.toJSON());
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        StateFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_state_free(ptr, 0);
    }
    /**
     * Retrieves the data associated with the state.
     *
     * # Returns
     * A `Data` instance containing the state data.
     * @returns {Data}
     */
    get data() {
        const ret = wasm.state_data(this.__wbg_ptr);
        return Data.__wrap(ret);
    }
    /**
     * Creates a new state from raw bytes.
     *
     * Use this for binary data like files, images, or serialized content.
     *
     * # Arguments
     * * `data` - The byte array data for the state.
     * * `metadata` - Optional metadata for the state.
     *
     * # Returns
     * A new `State` instance.
     * @param {Uint8Array} data
     * @param {string | null} [metadata]
     * @returns {State}
     */
    static fromBytes(data, metadata) {
        var ptr0 = isLikeNone(metadata) ? 0 : passStringToWasm0(metadata, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        const ret = wasm.state_fromBytes(data, ptr0, len0);
        return State.__wrap(ret);
    }
    /**
     * Creates a new state from a string.
     *
     * Use this for text data like documents, JSON, or configuration.
     *
     * # Arguments
     * * `data` - The string data for the state.
     * * `metadata` - Optional metadata for the state.
     *
     * # Returns
     * A new `State` instance.
     * @param {string} data
     * @param {string | null} [metadata]
     * @returns {State}
     */
    static fromString(data, metadata) {
        const ptr0 = passStringToWasm0(data, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        var ptr1 = isLikeNone(metadata) ? 0 : passStringToWasm0(metadata, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len1 = WASM_VECTOR_LEN;
        const ret = wasm.state_fromString(ptr0, len0, ptr1, len1);
        return State.__wrap(ret);
    }
    /**
     * Retrieves the metadata associated with the state.
     *
     * # Returns
     * A `string` containing the metadata, if existing.
     * @returns {string | undefined}
     */
    get metadata() {
        const ret = wasm.state_metadata(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
}
if (Symbol.dispose) State.prototype[Symbol.dispose] = State.prototype.free;
exports.State = State;

/**
 * Represents a time lock configuration.
 *
 * It allows the creation and inspection of time lock configurations for notarization objects.
 */
class TimeLock {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(TimeLock.prototype);
        obj.__wbg_ptr = ptr;
        TimeLockFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    toJSON() {
        return {
            args: this.args,
            type: this.type,
        };
    }
    toString() {
        return JSON.stringify(this);
    }
    [inspect.custom]() {
        return Object.assign(Object.create({constructor: this.constructor}), this.toJSON());
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        TimeLockFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_timelock_free(ptr, 0);
    }
    /**
     * Retrieves the arguments associated with the time lock.
     *
     * # Returns
     * An `any` value containing the arguments for the time lock:
     * - For `UnlockAt`, the timestamp is returned.
     * - For other types, `undefined` is returned.
     * @returns {any}
     */
    get args() {
        const ret = wasm.timelock_args(this.__wbg_ptr);
        return ret;
    }
    /**
     * Retrieves the type of the time lock.
     *
     * # Returns
     * The `TimeLockType` representing the type of the time lock.
     * @returns {TimeLockType}
     */
    get type() {
        const ret = wasm.timelock_type(this.__wbg_ptr);
        return __wbindgen_enum_TimeLockType[ret];
    }
    /**
     * Creates a time lock with no restrictions.
     *
     * # Returns
     * A new `TimeLock` instance with no time lock applied.
     * @returns {TimeLock}
     */
    static withNone() {
        const ret = wasm.timelock_withNone();
        return TimeLock.__wrap(ret);
    }
    /**
     * Creates a time lock that unlocks at a specific timestamp.
     *
     * # Arguments
     * * `time` - The timestamp in seconds since the Unix epoch at which the object will unlock.
     *
     * # Returns
     * A new `TimeLock` instance configured to unlock at the specified timestamp.
     * @param {number} time
     * @returns {TimeLock}
     */
    static withUnlockAt(time) {
        const ret = wasm.timelock_withUnlockAt(time);
        return TimeLock.__wrap(ret);
    }
    /**
     * Creates a time lock that remains locked until the object is destroyed.
     *
     * # Returns
     * A new `TimeLock` instance configured to remain locked until destruction.
     * @returns {TimeLock}
     */
    static withUntilDestroyed() {
        const ret = wasm.timelock_withUntilDestroyed();
        return TimeLock.__wrap(ret);
    }
}
if (Symbol.dispose) TimeLock.prototype[Symbol.dispose] = TimeLock.prototype.free;
exports.TimeLock = TimeLock;

class TransactionBuilder {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(TransactionBuilder.prototype);
        obj.__wbg_ptr = ptr;
        TransactionBuilderFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        TransactionBuilderFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_transactionbuilder_free(ptr, 0);
    }
    /**
     * @param {CoreClient} client
     * @returns {[Uint8Array, string[], Transaction]}
     */
    build(client) {
        const ptr = this.__destroy_into_raw();
        const ret = wasm.transactionbuilder_build(ptr, client);
        return ret;
    }
    /**
     * @param {CoreClient} client
     * @returns {TransactionOutput<unknown>}
     */
    buildAndExecute(client) {
        const ptr = this.__destroy_into_raw();
        const ret = wasm.transactionbuilder_buildAndExecute(ptr, client);
        return ret;
    }
    /**
     * @param {CoreClientReadOnly} client
     * @returns {[Uint8Array, string[], Transaction]}
     */
    build_with_defaults(client) {
        const ptr = this.__destroy_into_raw();
        const ret = wasm.transactionbuilder_build_with_defaults(ptr, client);
        return ret;
    }
    /**
     * Execute this transaction using an IOTA Gas Station.
     * @param {CoreClient} client
     * @param {string} gas_station_url
     * @param {HttpClient} http_client
     * @param {GasStationParams | null} [options]
     * @returns {Promise<TransactionOutput>}
     */
    executeWithGasStation(client, gas_station_url, http_client, options) {
        const ptr = this.__destroy_into_raw();
        const ptr0 = passStringToWasm0(gas_station_url, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        let ptr1 = 0;
        if (!isLikeNone(options)) {
            _assertClass(options, GasStationParams);
            ptr1 = options.__destroy_into_raw();
        }
        const ret = wasm.transactionbuilder_executeWithGasStation(ptr, client, ptr0, len0, http_client, ptr1);
        return ret;
    }
    /**
     * @param {Transaction<unknown>} tx
     */
    constructor(tx) {
        const ret = wasm.transactionbuilder_new(tx);
        this.__wbg_ptr = ret >>> 0;
        TransactionBuilderFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
    /**
     * @returns {Transaction<unknown>}
     */
    get transaction() {
        const ret = wasm.transactionbuilder_transaction(this.__wbg_ptr);
        return ret;
    }
    /**
     * @param {bigint} budget
     * @returns {TransactionBuilder}
     */
    withGasBudget(budget) {
        const ptr = this.__destroy_into_raw();
        const ret = wasm.transactionbuilder_withGasBudget(ptr, budget);
        return TransactionBuilder.__wrap(ret);
    }
    /**
     * @param {string} owner
     * @returns {TransactionBuilder}
     */
    withGasOwner(owner) {
        const ptr = this.__destroy_into_raw();
        const ptr0 = passStringToWasm0(owner, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.transactionbuilder_withGasOwner(ptr, ptr0, len0);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return TransactionBuilder.__wrap(ret[0]);
    }
    /**
     * @param {IotaObjectRef[]} payment
     * @returns {TransactionBuilder}
     */
    withGasPayment(payment) {
        const ptr = this.__destroy_into_raw();
        const ptr0 = passArrayJsValueToWasm0(payment, wasm.__wbindgen_malloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.transactionbuilder_withGasPayment(ptr, ptr0, len0);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return TransactionBuilder.__wrap(ret[0]);
    }
    /**
     * @param {bigint} price
     * @returns {TransactionBuilder}
     */
    withGasPrice(price) {
        const ptr = this.__destroy_into_raw();
        const ret = wasm.transactionbuilder_withGasPrice(ptr, price);
        return TransactionBuilder.__wrap(ret);
    }
    /**
     * @param {string} sender
     * @returns {TransactionBuilder}
     */
    withSender(sender) {
        const ptr = this.__destroy_into_raw();
        const ptr0 = passStringToWasm0(sender, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.transactionbuilder_withSender(ptr, ptr0, len0);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return TransactionBuilder.__wrap(ret[0]);
    }
    /**
     * @param {CoreClient} client
     * @returns {Promise<TransactionBuilder>}
     */
    withSignature(client) {
        const ptr = this.__destroy_into_raw();
        const ret = wasm.transactionbuilder_withSignature(ptr, client);
        return ret;
    }
    /**
     * @param {CoreClientReadOnly} client
     * @param {SponsorFn} sponsor_fn
     * @returns {Promise<TransactionBuilder>}
     */
    withSponsor(client, sponsor_fn) {
        const ptr = this.__destroy_into_raw();
        const ret = wasm.transactionbuilder_withSponsor(ptr, client, sponsor_fn);
        return ret;
    }
}
if (Symbol.dispose) TransactionBuilder.prototype[Symbol.dispose] = TransactionBuilder.prototype.free;
exports.TransactionBuilder = TransactionBuilder;

class TransactionOutput {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(TransactionOutput.prototype);
        obj.__wbg_ptr = ptr;
        TransactionOutputFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    toJSON() {
        return {
            output: this.output,
            response: this.response,
        };
    }
    toString() {
        return JSON.stringify(this);
    }
    [inspect.custom]() {
        return Object.assign(Object.create({constructor: this.constructor}), this.toJSON());
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        TransactionOutputFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_transactionoutput_free(ptr, 0);
    }
    /**
     * @returns {any}
     */
    get output() {
        const ret = wasm.__wbg_get_transactionoutput_output(this.__wbg_ptr);
        return ret;
    }
    /**
     * @returns {IotaTransactionBlockResponse}
     */
    get response() {
        const ret = wasm.__wbg_get_transactionoutput_response(this.__wbg_ptr);
        return ret;
    }
    /**
     * @param {any} arg0
     */
    set output(arg0) {
        wasm.__wbg_set_transactionoutput_output(this.__wbg_ptr, arg0);
    }
    /**
     * @param {IotaTransactionBlockResponse} arg0
     */
    set response(arg0) {
        wasm.__wbg_set_transactionoutput_response(this.__wbg_ptr, arg0);
    }
}
if (Symbol.dispose) TransactionOutput.prototype[Symbol.dispose] = TransactionOutput.prototype.free;
exports.TransactionOutput = TransactionOutput;

/**
 * Represents a transaction for transferring a dynamic notarization to a new owner.
 *
 * This is only available for dynamic notarization's
 */
class TransferNotarization {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(TransferNotarization.prototype);
        obj.__wbg_ptr = ptr;
        TransferNotarizationFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    toJSON() {
        return {
        };
    }
    toString() {
        return JSON.stringify(this);
    }
    [inspect.custom]() {
        return Object.assign(Object.create({constructor: this.constructor}), this.toJSON());
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        TransferNotarizationFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_transfernotarization_free(ptr, 0);
    }
    /**
     * Applies transaction effects and events to this transfer operation.
     *
     * This method is called automatically by Transaction::build_programmable_transaction()
     * and Transaction::apply() methods after the transaction has been successfully submitted
     * to process the results from the ledger.
     *
     * # Arguments
     * * `effects` - The transaction block effects to apply.
     * * `events` - The transaction block events to apply.
     * @param {TransactionEffects} wasm_effects
     * @param {IotaEvent[]} wasm_events
     * @param {CoreClientReadOnly} client
     * @returns {Promise<Empty>}
     */
    applyWithEvents(wasm_effects, wasm_events, client) {
        const ptr = this.__destroy_into_raw();
        const ret = wasm.transfernotarization_applyWithEvents(ptr, wasm_effects, wasm_events, client);
        return ret;
    }
    /**
     * Builds and returns a programmable transaction for transferring a notarization.
     *
     * # Returns
     * The binary BCS serialization of the programmable transaction.
     * This transaction can be submitted to the network to transfer a notarization.
     *
     * # Errors
     * Returns an error if the transaction cannot be built due to invalid parameters
     * or other constraints.
     * @param {CoreClientReadOnly} client
     * @returns {Promise<Uint8Array>}
     */
    buildProgrammableTransaction(client) {
        const ret = wasm.transfernotarization_buildProgrammableTransaction(this.__wbg_ptr, client);
        return ret;
    }
    /**
     * @param {string} recipient
     * @param {string} object_id
     */
    constructor(recipient, object_id) {
        const ptr0 = passStringToWasm0(recipient, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ptr1 = passStringToWasm0(object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len1 = WASM_VECTOR_LEN;
        const ret = wasm.transfernotarization_new(ptr0, len0, ptr1, len1);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        this.__wbg_ptr = ret[0] >>> 0;
        TransferNotarizationFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
}
if (Symbol.dispose) TransferNotarization.prototype[Symbol.dispose] = TransferNotarization.prototype.free;
exports.TransferNotarization = TransferNotarization;

/**
 * Represents a transaction for updating the metadata of a notarization.
 */
class UpdateMetadata {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(UpdateMetadata.prototype);
        obj.__wbg_ptr = ptr;
        UpdateMetadataFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    toJSON() {
        return {
        };
    }
    toString() {
        return JSON.stringify(this);
    }
    [inspect.custom]() {
        return Object.assign(Object.create({constructor: this.constructor}), this.toJSON());
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        UpdateMetadataFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_updatemetadata_free(ptr, 0);
    }
    /**
     * Applies transaction effects and events to this metadata update operation.
     *
     * This method is called automatically by Transaction::build_programmable_transaction()
     * and Transaction::apply() methods after the transaction has been successfully submitted
     * to process the results from the ledger.
     *
     * # Arguments
     * * `effects` - The transaction block effects to apply.
     * * `events` - The transaction block events to apply.
     * @param {TransactionEffects} wasm_effects
     * @param {IotaEvent[]} wasm_events
     * @param {CoreClientReadOnly} client
     * @returns {Promise<Empty>}
     */
    applyWithEvents(wasm_effects, wasm_events, client) {
        const ptr = this.__destroy_into_raw();
        const ret = wasm.updatemetadata_applyWithEvents(ptr, wasm_effects, wasm_events, client);
        return ret;
    }
    /**
     * Builds and returns a programmable transaction for updating the metadata of a notarization.
     *
     * # Returns
     * The binary BCS serialization of the programmable transaction.
     * This transaction can be submitted to the network to update the metadata of a notarization.
     *
     * # Errors
     * Returns an error if the transaction cannot be built due to invalid parameters
     * or other constraints.
     * @param {CoreClientReadOnly} client
     * @returns {Promise<Uint8Array>}
     */
    buildProgrammableTransaction(client) {
        const ret = wasm.updatemetadata_buildProgrammableTransaction(this.__wbg_ptr, client);
        return ret;
    }
    /**
     * @param {string | null | undefined} metadata
     * @param {string} object_id
     */
    constructor(metadata, object_id) {
        var ptr0 = isLikeNone(metadata) ? 0 : passStringToWasm0(metadata, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        var len0 = WASM_VECTOR_LEN;
        const ptr1 = passStringToWasm0(object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len1 = WASM_VECTOR_LEN;
        const ret = wasm.updatemetadata_new(ptr0, len0, ptr1, len1);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        this.__wbg_ptr = ret[0] >>> 0;
        UpdateMetadataFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
}
if (Symbol.dispose) UpdateMetadata.prototype[Symbol.dispose] = UpdateMetadata.prototype.free;
exports.UpdateMetadata = UpdateMetadata;

/**
 * Represents a transaction for updating the state of a dynamic notarization.
 *
 * This is only available for dynamic notarization's
 */
class UpdateState {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(UpdateState.prototype);
        obj.__wbg_ptr = ptr;
        UpdateStateFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    toJSON() {
        return {
        };
    }
    toString() {
        return JSON.stringify(this);
    }
    [inspect.custom]() {
        return Object.assign(Object.create({constructor: this.constructor}), this.toJSON());
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        UpdateStateFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_updatestate_free(ptr, 0);
    }
    /**
     * Applies transaction effects and events to this state update operation.
     *
     * This method is called automatically by Transaction::build_programmable_transaction()
     * and Transaction::apply() methods after the transaction has been successfully submitted
     * to process the results from the ledger.
     *
     * # Arguments
     * * `effects` - The transaction block effects to apply.
     * * `events` - The transaction block events to apply.
     * @param {TransactionEffects} wasm_effects
     * @param {IotaEvent[]} wasm_events
     * @param {CoreClientReadOnly} client
     * @returns {Promise<Empty>}
     */
    applyWithEvents(wasm_effects, wasm_events, client) {
        const ptr = this.__destroy_into_raw();
        const ret = wasm.updatestate_applyWithEvents(ptr, wasm_effects, wasm_events, client);
        return ret;
    }
    /**
     * Builds and returns a programmable transaction for updating the state of a notarization.
     *
     * # Returns
     * The binary BCS serialization of the programmable transaction.
     * This transaction can be submitted to the network to updating the state of a notarization.
     *
     * # Errors
     * Returns an error if the transaction cannot be built due to invalid parameters
     * or other constraints.
     * @param {CoreClientReadOnly} client
     * @returns {Promise<Uint8Array>}
     */
    buildProgrammableTransaction(client) {
        const ret = wasm.updatestate_buildProgrammableTransaction(this.__wbg_ptr, client);
        return ret;
    }
    /**
     * @param {State} state
     * @param {string} object_id
     */
    constructor(state, object_id) {
        _assertClass(state, State);
        var ptr0 = state.__destroy_into_raw();
        const ptr1 = passStringToWasm0(object_id, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len1 = WASM_VECTOR_LEN;
        const ret = wasm.updatestate_new(ptr0, ptr1, len1);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        this.__wbg_ptr = ret[0] >>> 0;
        UpdateStateFinalization.register(this, this.__wbg_ptr, this);
        return this;
    }
}
if (Symbol.dispose) UpdateState.prototype[Symbol.dispose] = UpdateState.prototype.free;
exports.UpdateState = UpdateState;

class WasmManagedCoreClient {
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        WasmManagedCoreClientFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_wasmmanagedcoreclient_free(ptr, 0);
    }
    /**
     * @returns {IotaClient}
     */
    iotaClient() {
        const ret = wasm.wasmmanagedcoreclient_iotaClient(this.__wbg_ptr);
        return ret;
    }
    /**
     * @returns {string}
     */
    network() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.wasmmanagedcoreclient_network(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * @returns {string[]}
     */
    packageHistory() {
        const ret = wasm.wasmmanagedcoreclient_packageHistory(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * @returns {string}
     */
    packageId() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.wasmmanagedcoreclient_packageId(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * @returns {string}
     */
    senderAddress() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.wasmmanagedcoreclient_senderAddress(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * @returns {PublicKey}
     */
    senderPublicKey() {
        const ret = wasm.wasmmanagedcoreclient_senderPublicKey(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return takeFromExternrefTable0(ret[0]);
    }
    /**
     * @returns {TransactionSigner}
     */
    signer() {
        const ret = wasm.wasmmanagedcoreclient_signer(this.__wbg_ptr);
        return ret;
    }
}
if (Symbol.dispose) WasmManagedCoreClient.prototype[Symbol.dispose] = WasmManagedCoreClient.prototype.free;
exports.WasmManagedCoreClient = WasmManagedCoreClient;

class WasmManagedCoreClientReadOnly {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(WasmManagedCoreClientReadOnly.prototype);
        obj.__wbg_ptr = ptr;
        WasmManagedCoreClientReadOnlyFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        WasmManagedCoreClientReadOnlyFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_wasmmanagedcoreclientreadonly_free(ptr, 0);
    }
    /**
     * @returns {IotaClient}
     */
    iotaClient() {
        const ret = wasm.wasmmanagedcoreclientreadonly_iotaClient(this.__wbg_ptr);
        return ret;
    }
    /**
     * @returns {string}
     */
    network() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.wasmmanagedcoreclientreadonly_network(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * @returns {string[]}
     */
    packageHistory() {
        const ret = wasm.wasmmanagedcoreclientreadonly_packageHistory(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * @returns {string}
     */
    packageId() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.wasmmanagedcoreclientreadonly_packageId(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        } finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
}
if (Symbol.dispose) WasmManagedCoreClientReadOnly.prototype[Symbol.dispose] = WasmManagedCoreClientReadOnly.prototype.free;
exports.WasmManagedCoreClientReadOnly = WasmManagedCoreClientReadOnly;

/**
 * Initializes the console error panic hook for better error messages
 */
function start() {
    wasm.start();
}
exports.start = start;

function __wbg_get_imports() {
    const import0 = {
        __proto__: null,
        __wbg_Error_8c4e43fe74559d73: function(arg0, arg1) {
            const ret = Error(getStringFromWasm0(arg0, arg1));
            return ret;
        },
        __wbg_Number_04624de7d0e8332d: function(arg0) {
            const ret = Number(arg0);
            return ret;
        },
        __wbg_String_8f0eb39a4a4c2f66: function(arg0, arg1) {
            const ret = String(arg1);
            const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            const len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg___wbindgen_bigint_get_as_i64_8fcf4ce7f1ca72a2: function(arg0, arg1) {
            const v = arg1;
            const ret = typeof(v) === 'bigint' ? v : undefined;
            getDataViewMemory0().setBigInt64(arg0 + 8 * 1, isLikeNone(ret) ? BigInt(0) : ret, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, !isLikeNone(ret), true);
        },
        __wbg___wbindgen_boolean_get_bbbb1c18aa2f5e25: function(arg0) {
            const v = arg0;
            const ret = typeof(v) === 'boolean' ? v : undefined;
            return isLikeNone(ret) ? 0xFFFFFF : ret ? 1 : 0;
        },
        __wbg___wbindgen_debug_string_0bc8482c6e3508ae: function(arg0, arg1) {
            const ret = debugString(arg1);
            const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            const len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg___wbindgen_in_47fa6863be6f2f25: function(arg0, arg1) {
            const ret = arg0 in arg1;
            return ret;
        },
        __wbg___wbindgen_is_bigint_31b12575b56f32fc: function(arg0) {
            const ret = typeof(arg0) === 'bigint';
            return ret;
        },
        __wbg___wbindgen_is_function_0095a73b8b156f76: function(arg0) {
            const ret = typeof(arg0) === 'function';
            return ret;
        },
        __wbg___wbindgen_is_object_5ae8e5880f2c1fbd: function(arg0) {
            const val = arg0;
            const ret = typeof(val) === 'object' && val !== null;
            return ret;
        },
        __wbg___wbindgen_is_string_cd444516edc5b180: function(arg0) {
            const ret = typeof(arg0) === 'string';
            return ret;
        },
        __wbg___wbindgen_is_undefined_9e4d92534c42d778: function(arg0) {
            const ret = arg0 === undefined;
            return ret;
        },
        __wbg___wbindgen_jsval_eq_11888390b0186270: function(arg0, arg1) {
            const ret = arg0 === arg1;
            return ret;
        },
        __wbg___wbindgen_jsval_loose_eq_9dd77d8cd6671811: function(arg0, arg1) {
            const ret = arg0 == arg1;
            return ret;
        },
        __wbg___wbindgen_number_get_8ff4255516ccad3e: function(arg0, arg1) {
            const obj = arg1;
            const ret = typeof(obj) === 'number' ? obj : undefined;
            getDataViewMemory0().setFloat64(arg0 + 8 * 1, isLikeNone(ret) ? 0 : ret, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, !isLikeNone(ret), true);
        },
        __wbg___wbindgen_string_get_72fb696202c56729: function(arg0, arg1) {
            const obj = arg1;
            const ret = typeof(obj) === 'string' ? obj : undefined;
            var ptr1 = isLikeNone(ret) ? 0 : passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            var len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg___wbindgen_throw_be289d5034ed271b: function(arg0, arg1) {
            throw new Error(getStringFromWasm0(arg0, arg1));
        },
        __wbg__wbg_cb_unref_d9b87ff7982e3b21: function(arg0) {
            arg0._wbg_cb_unref();
        },
        __wbg_abort_2f0584e03e8e3950: function(arg0) {
            arg0.abort();
        },
        __wbg_abort_d549b92d3c665de1: function(arg0, arg1) {
            arg0.abort(arg1);
        },
        __wbg_append_a992ccc37aa62dc4: function() { return handleError(function (arg0, arg1, arg2, arg3, arg4) {
            arg0.append(getStringFromWasm0(arg1, arg2), getStringFromWasm0(arg3, arg4));
        }, arguments); },
        __wbg_applyWithEvents_522768b04c7228b8: function() { return handleError(function (arg0, arg1, arg2, arg3) {
            const ret = arg0.applyWithEvents(arg1, arg2, arg3);
            return ret;
        }, arguments); },
        __wbg_apply_a89507ce04e01050: function() { return handleError(function (arg0, arg1, arg2) {
            const ret = arg0.apply(arg1, arg2);
            return ret;
        }, arguments); },
        __wbg_arrayBuffer_bb54076166006c39: function() { return handleError(function (arg0) {
            const ret = arg0.arrayBuffer();
            return ret;
        }, arguments); },
        __wbg_assign_6170c0d04d5c26f4: function(arg0, arg1) {
            const ret = Object.assign(arg0, arg1);
            return ret;
        },
        __wbg_buildProgrammableTransaction_ef93d4f5818528f7: function() { return handleError(function (arg0, arg1) {
            const ret = arg0.buildProgrammableTransaction(arg1);
            return ret;
        }, arguments); },
        __wbg_build_ce1320a6d080a98e: function() { return handleError(function (arg0) {
            const ret = arg0.build();
            return ret;
        }, arguments); },
        __wbg_call_389efe28435a9388: function() { return handleError(function (arg0, arg1) {
            const ret = arg0.call(arg1);
            return ret;
        }, arguments); },
        __wbg_call_4708e0c13bdc8e95: function() { return handleError(function (arg0, arg1, arg2) {
            const ret = arg0.call(arg1, arg2);
            return ret;
        }, arguments); },
        __wbg_clearTimeout_26e350acd8252ec6: function(arg0) {
            const ret = clearTimeout(arg0);
            return ret;
        },
        __wbg_createnotarizationdynamic_new: function(arg0) {
            const ret = CreateNotarizationDynamic.__wrap(arg0);
            return ret;
        },
        __wbg_createnotarizationlocked_new: function(arg0) {
            const ret = CreateNotarizationLocked.__wrap(arg0);
            return ret;
        },
        __wbg_destroynotarization_new: function(arg0) {
            const ret = DestroyNotarization.__wrap(arg0);
            return ret;
        },
        __wbg_devInspectTransactionBlock_24ccdd29024f7195: function(arg0, arg1) {
            const ret = arg0.devInspectTransactionBlock(arg1);
            return ret;
        },
        __wbg_digest_b523852747c4f0ca: function(arg0, arg1) {
            const ret = arg1.digest;
            const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            const len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg_done_57b39ecd9addfe81: function(arg0) {
            const ret = arg0.done;
            return ret;
        },
        __wbg_empty_new: function(arg0) {
            const ret = Empty.__wrap(arg0);
            return ret;
        },
        __wbg_entries_04679661ea6e74fc: function(arg0) {
            const ret = arg0.entries();
            return ret;
        },
        __wbg_entries_58c7934c745daac7: function(arg0) {
            const ret = Object.entries(arg0);
            return ret;
        },
        __wbg_error_7534b8e9a36f1ab4: function(arg0, arg1) {
            let deferred0_0;
            let deferred0_1;
            try {
                deferred0_0 = arg0;
                deferred0_1 = arg1;
                console.error(getStringFromWasm0(arg0, arg1));
            } finally {
                wasm.__wbindgen_free(deferred0_0, deferred0_1, 1);
            }
        },
        __wbg_executeTransactionBlock_fa77a08f7ea531f1: function(arg0, arg1) {
            const ret = arg0.executeTransactionBlock(arg1);
            return ret;
        },
        __wbg_fetch_995a2faca6a97afe: function(arg0) {
            const ret = fetch(arg0);
            return ret;
        },
        __wbg_fetch_afb6a4b6cacf876d: function(arg0, arg1) {
            const ret = arg0.fetch(arg1);
            return ret;
        },
        __wbg_fromBytes_dbb4a8349f1524a3: function() { return handleError(function (arg0) {
            const ret = TransactionDataBuilder.fromBytes(arg0);
            return ret;
        }, arguments); },
        __wbg_gasstationparams_new: function(arg0) {
            const ret = GasStationParams.__wrap(arg0);
            return ret;
        },
        __wbg_getChainIdentifier_e269f70ad07327f8: function(arg0) {
            const ret = arg0.getChainIdentifier();
            return ret;
        },
        __wbg_getCoins_57e0426f0df3a6b6: function(arg0, arg1) {
            const ret = arg0.getCoins(arg1);
            return ret;
        },
        __wbg_getDynamicFieldObjectV2_44237aa296dafe22: function(arg0, arg1) {
            const ret = arg0.getDynamicFieldObjectV2(arg1);
            return ret;
        },
        __wbg_getDynamicFieldObject_3e3b457de8967680: function(arg0, arg1) {
            const ret = arg0.getDynamicFieldObject(arg1);
            return ret;
        },
        __wbg_getObject_4cf518c8c622cff7: function(arg0, arg1) {
            const ret = arg0.getObject(arg1);
            return ret;
        },
        __wbg_getOwnedObjects_ab05b1666074e2d8: function(arg0, arg1) {
            const ret = arg0.getOwnedObjects(arg1);
            return ret;
        },
        __wbg_getReferenceGasPrice_ecc52d6d0de7ef2a: function(arg0) {
            const ret = arg0.getReferenceGasPrice();
            return ret;
        },
        __wbg_getTransactionBlock_2006b397d4713946: function(arg0, arg1) {
            const ret = arg0.getTransactionBlock(arg1);
            return ret;
        },
        __wbg_get_9b94d73e6221f75c: function(arg0, arg1) {
            const ret = arg0[arg1 >>> 0];
            return ret;
        },
        __wbg_get_b3ed3ad4be2bc8ac: function() { return handleError(function (arg0, arg1) {
            const ret = Reflect.get(arg0, arg1);
            return ret;
        }, arguments); },
        __wbg_get_digest_61eb5c47a66a3e85: function(arg0, arg1) {
            const ret = arg1.get_digest();
            const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            const len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg_get_effects_df4a9afe9c280202: function(arg0) {
            const ret = arg0.get_effects();
            return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
        },
        __wbg_get_events_8484e0a54caf896b: function(arg0) {
            const ret = arg0.get_events();
            return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
        },
        __wbg_get_response_57b827b697d07aea: function(arg0) {
            const ret = arg0.get_response();
            return ret;
        },
        __wbg_get_with_ref_key_1dc361bd10053bfe: function(arg0, arg1) {
            const ret = arg0[arg1];
            return ret;
        },
        __wbg_has_d4e53238966c12b6: function() { return handleError(function (arg0, arg1) {
            const ret = Reflect.has(arg0, arg1);
            return ret;
        }, arguments); },
        __wbg_headers_59a2938db9f80985: function(arg0) {
            const ret = arg0.headers;
            return ret;
        },
        __wbg_instanceof_ArrayBuffer_c367199e2fa2aa04: function(arg0) {
            let result;
            try {
                result = arg0 instanceof ArrayBuffer;
            } catch (_) {
                result = false;
            }
            const ret = result;
            return ret;
        },
        __wbg_instanceof_Error_8573fe0b0b480f46: function(arg0) {
            let result;
            try {
                result = arg0 instanceof Error;
            } catch (_) {
                result = false;
            }
            const ret = result;
            return ret;
        },
        __wbg_instanceof_Map_53af74335dec57f4: function(arg0) {
            let result;
            try {
                result = arg0 instanceof Map;
            } catch (_) {
                result = false;
            }
            const ret = result;
            return ret;
        },
        __wbg_instanceof_Promise_0094681e3519d6ec: function(arg0) {
            let result;
            try {
                result = arg0 instanceof Promise;
            } catch (_) {
                result = false;
            }
            const ret = result;
            return ret;
        },
        __wbg_instanceof_Response_ee1d54d79ae41977: function(arg0) {
            let result;
            try {
                result = arg0 instanceof Response;
            } catch (_) {
                result = false;
            }
            const ret = result;
            return ret;
        },
        __wbg_instanceof_Uint8Array_9b9075935c74707c: function(arg0) {
            let result;
            try {
                result = arg0 instanceof Uint8Array;
            } catch (_) {
                result = false;
            }
            const ret = result;
            return ret;
        },
        __wbg_iotaClient_042b852656a5ec59: function(arg0) {
            const ret = arg0.iotaClient();
            return ret;
        },
        __wbg_iotaPublicKeyBytes_cc42b5652d1b6d6c: function() { return handleError(function (arg0) {
            const ret = arg0.iotaPublicKeyBytes();
            return ret;
        }, arguments); },
        __wbg_isArray_d314bb98fcf08331: function(arg0) {
            const ret = Array.isArray(arg0);
            return ret;
        },
        __wbg_isSafeInteger_bfbc7332a9768d2a: function(arg0) {
            const ret = Number.isSafeInteger(arg0);
            return ret;
        },
        __wbg_iterator_6ff6560ca1568e55: function() {
            const ret = Symbol.iterator;
            return ret;
        },
        __wbg_length_32ed9a279acd054c: function(arg0) {
            const ret = arg0.length;
            return ret;
        },
        __wbg_length_35a7bace40f36eac: function(arg0) {
            const ret = arg0.length;
            return ret;
        },
        __wbg_lockmetadata_new: function(arg0) {
            const ret = LockMetadata.__wrap(arg0);
            return ret;
        },
        __wbg_log_9a40138eaafa73b7: function(arg0, arg1) {
            console.log(getStringFromWasm0(arg0, arg1));
        },
        __wbg_message_9ddc4b9a62a7c379: function(arg0) {
            const ret = arg0.message;
            return ret;
        },
        __wbg_name_446e25ef2cfdab5a: function(arg0) {
            const ret = arg0.name;
            return ret;
        },
        __wbg_network_bba2c2a26576607c: function(arg0, arg1) {
            const ret = arg1.network();
            const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            const len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg_new_361308b2356cecd0: function() {
            const ret = new Object();
            return ret;
        },
        __wbg_new_3eb36ae241fe6f44: function() {
            const ret = new Array();
            return ret;
        },
        __wbg_new_412c5debaae4d54b: function(arg0) {
            const ret = new WasmIotaTransactionBlockResponseWrapper(arg0);
            return ret;
        },
        __wbg_new_64284bd487f9d239: function() { return handleError(function () {
            const ret = new Headers();
            return ret;
        }, arguments); },
        __wbg_new_72b49615380db768: function(arg0, arg1) {
            const ret = new Error(getStringFromWasm0(arg0, arg1));
            return ret;
        },
        __wbg_new_8a6f238a6ece86ea: function() {
            const ret = new Error();
            return ret;
        },
        __wbg_new_b5d9e2fb389fef91: function(arg0, arg1) {
            try {
                var state0 = {a: arg0, b: arg1};
                var cb0 = (arg0, arg1) => {
                    const a = state0.a;
                    state0.a = 0;
                    try {
                        return wasm_bindgen__convert__closures_____invoke__h9ec99707e3310bed(a, state0.b, arg0, arg1);
                    } finally {
                        state0.a = a;
                    }
                };
                const ret = new Promise(cb0);
                return ret;
            } finally {
                state0.a = state0.b = 0;
            }
        },
        __wbg_new_b949e7f56150a5d1: function() { return handleError(function () {
            const ret = new AbortController();
            return ret;
        }, arguments); },
        __wbg_new_dca287b076112a51: function() {
            const ret = new Map();
            return ret;
        },
        __wbg_new_dd2b680c8bf6ae29: function(arg0) {
            const ret = new Uint8Array(arg0);
            return ret;
        },
        __wbg_new_ed25519_pk_base64_07d0234a26acd375: function() { return handleError(function (arg0, arg1) {
            const ret = new Ed25519PublicKey(getStringFromWasm0(arg0, arg1));
            return ret;
        }, arguments); },
        __wbg_new_from_slice_a3d2629dc1826784: function(arg0, arg1) {
            const ret = new Uint8Array(getArrayU8FromWasm0(arg0, arg1));
            return ret;
        },
        __wbg_new_no_args_1c7c842f08d00ebb: function(arg0, arg1) {
            const ret = new Function(getStringFromWasm0(arg0, arg1));
            return ret;
        },
        __wbg_new_secp256k1_pk_base64_f67770d6dbe04680: function() { return handleError(function (arg0, arg1) {
            const ret = new Secp256k1PublicKey(getStringFromWasm0(arg0, arg1));
            return ret;
        }, arguments); },
        __wbg_new_secp256r1_pk_base64_8de106561d539729: function() { return handleError(function (arg0, arg1) {
            const ret = new Secp256r1PublicKey(getStringFromWasm0(arg0, arg1));
            return ret;
        }, arguments); },
        __wbg_new_with_str_and_init_a61cbc6bdef21614: function() { return handleError(function (arg0, arg1, arg2) {
            const ret = new Request(getStringFromWasm0(arg0, arg1), arg2);
            return ret;
        }, arguments); },
        __wbg_next_3482f54c49e8af19: function() { return handleError(function (arg0) {
            const ret = arg0.next();
            return ret;
        }, arguments); },
        __wbg_next_418f80d8f5303233: function(arg0) {
            const ret = arg0.next;
            return ret;
        },
        __wbg_notarizationclient_new: function(arg0) {
            const ret = NotarizationClient.__wrap(arg0);
            return ret;
        },
        __wbg_notarizationclientreadonly_new: function(arg0) {
            const ret = NotarizationClientReadOnly.__wrap(arg0);
            return ret;
        },
        __wbg_objectId_4e8003ff8c07f0c8: function(arg0, arg1) {
            const ret = arg1.objectId;
            const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            const len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg_onchainnotarization_new: function(arg0) {
            const ret = OnChainNotarization.__wrap(arg0);
            return ret;
        },
        __wbg_packageHistory_9948ba04d617b46f: function(arg0, arg1) {
            const ret = arg1.packageHistory();
            const ptr1 = passArrayJsValueToWasm0(ret, wasm.__wbindgen_malloc);
            const len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg_prototypesetcall_bdcdcc5842e4d77d: function(arg0, arg1, arg2) {
            Uint8Array.prototype.set.call(getArrayU8FromWasm0(arg0, arg1), arg2);
        },
        __wbg_push_8ffdcb2063340ba5: function(arg0, arg1) {
            const ret = arg0.push(arg1);
            return ret;
        },
        __wbg_queueMicrotask_0aa0a927f78f5d98: function(arg0) {
            const ret = arg0.queueMicrotask;
            return ret;
        },
        __wbg_queueMicrotask_5bb536982f78a56f: function(arg0) {
            queueMicrotask(arg0);
        },
        __wbg_resolve_002c4b7d9d8f6b64: function(arg0) {
            const ret = Promise.resolve(arg0);
            return ret;
        },
        __wbg_send_d397938be2c37991: function() { return handleError(function (arg0, arg1) {
            const ret = arg0.send(arg1);
            return ret;
        }, arguments); },
        __wbg_senderAddress_bdfc25805f63810a: function(arg0, arg1) {
            const ret = arg1.senderAddress();
            const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            const len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg_senderPublicKey_59cc238f40a9e12c: function(arg0) {
            const ret = arg0.senderPublicKey();
            return ret;
        },
        __wbg_setTimeout_4360465139fa5dfe: function(arg0, arg1) {
            const ret = setTimeout(arg0, arg1);
            return ret;
        },
        __wbg_set_1eb0999cf5d27fc8: function(arg0, arg1, arg2) {
            const ret = arg0.set(arg1, arg2);
            return ret;
        },
        __wbg_set_3f1d0b984ed272ed: function(arg0, arg1, arg2) {
            arg0[arg1] = arg2;
        },
        __wbg_set_body_9a7e00afe3cfe244: function(arg0, arg1) {
            arg0.body = arg1;
        },
        __wbg_set_cache_315a3ed773a41543: function(arg0, arg1) {
            arg0.cache = __wbindgen_enum_RequestCache[arg1];
        },
        __wbg_set_credentials_c4a58d2e05ef24fb: function(arg0, arg1) {
            arg0.credentials = __wbindgen_enum_RequestCredentials[arg1];
        },
        __wbg_set_f43e577aea94465b: function(arg0, arg1, arg2) {
            arg0[arg1 >>> 0] = arg2;
        },
        __wbg_set_headers_cfc5f4b2c1f20549: function(arg0, arg1) {
            arg0.headers = arg1;
        },
        __wbg_set_method_c3e20375f5ae7fac: function(arg0, arg1, arg2) {
            arg0.method = getStringFromWasm0(arg1, arg2);
        },
        __wbg_set_mode_b13642c312648202: function(arg0, arg1) {
            arg0.mode = __wbindgen_enum_RequestMode[arg1];
        },
        __wbg_set_name_2aca19de8d382675: function(arg0, arg1, arg2) {
            arg0.name = getStringFromWasm0(arg1, arg2);
        },
        __wbg_set_signal_f2d3f8599248896d: function(arg0, arg1) {
            arg0.signal = arg1;
        },
        __wbg_sign_b8f4eaf436fd96b5: function() { return handleError(function (arg0, arg1, arg2) {
            var v0 = getArrayU8FromWasm0(arg1, arg2).slice();
            wasm.__wbindgen_free(arg1, arg2 * 1, 1);
            const ret = arg0.sign(v0);
            return ret;
        }, arguments); },
        __wbg_signal_d1285ecab4ebc5ad: function(arg0) {
            const ret = arg0.signal;
            return ret;
        },
        __wbg_signer_074238a50af388df: function(arg0) {
            const ret = arg0.signer();
            return ret;
        },
        __wbg_stack_0ed75d68575b0f3c: function(arg0, arg1) {
            const ret = arg1.stack;
            const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            const len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg_state_new: function(arg0) {
            const ret = State.__wrap(arg0);
            return ret;
        },
        __wbg_static_accessor_GLOBAL_12837167ad935116: function() {
            const ret = typeof global === 'undefined' ? null : global;
            return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
        },
        __wbg_static_accessor_GLOBAL_THIS_e628e89ab3b1c95f: function() {
            const ret = typeof globalThis === 'undefined' ? null : globalThis;
            return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
        },
        __wbg_static_accessor_SELF_a621d3dfbb60d0ce: function() {
            const ret = typeof self === 'undefined' ? null : self;
            return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
        },
        __wbg_static_accessor_WINDOW_f8727f0cf888e0bd: function() {
            const ret = typeof window === 'undefined' ? null : window;
            return isLikeNone(ret) ? 0 : addToExternrefTable0(ret);
        },
        __wbg_status_89d7e803db911ee7: function(arg0) {
            const ret = arg0.status;
            return ret;
        },
        __wbg_stringify_e4a940b133e6b7d8: function(arg0, arg1) {
            const ret = JSON.stringify(arg1);
            var ptr1 = isLikeNone(ret) ? 0 : passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            var len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg_then_0d9fe2c7b1857d32: function(arg0, arg1, arg2) {
            const ret = arg0.then(arg1, arg2);
            return ret;
        },
        __wbg_then_b9e7b3b5f1a9e1b5: function(arg0, arg1) {
            const ret = arg0.then(arg1);
            return ret;
        },
        __wbg_toIotaPublicKey_040e1e58e8c9a288: function(arg0, arg1) {
            const ret = arg1.toIotaPublicKey();
            const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            const len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg_toString_029ac24421fd7a24: function(arg0) {
            const ret = arg0.toString();
            return ret;
        },
        __wbg_to_string_5d9975e5ee91a211: function(arg0, arg1) {
            const ret = arg1.to_string();
            const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            const len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg_transactionbuilder_new: function(arg0) {
            const ret = TransactionBuilder.__wrap(arg0);
            return ret;
        },
        __wbg_transactionoutput_new: function(arg0) {
            const ret = TransactionOutput.__wrap(arg0);
            return ret;
        },
        __wbg_transfernotarization_new: function(arg0) {
            const ret = TransferNotarization.__wrap(arg0);
            return ret;
        },
        __wbg_updatemetadata_new: function(arg0) {
            const ret = UpdateMetadata.__wrap(arg0);
            return ret;
        },
        __wbg_updatestate_new: function(arg0) {
            const ret = UpdateState.__wrap(arg0);
            return ret;
        },
        __wbg_url_c484c26b1fbf5126: function(arg0, arg1) {
            const ret = arg1.url;
            const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            const len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg_value_0546255b415e96c1: function(arg0) {
            const ret = arg0.value;
            return ret;
        },
        __wbg_version_4e1ee1eb213cea66: function(arg0, arg1) {
            const ret = arg1.version;
            const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
            const len1 = WASM_VECTOR_LEN;
            getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
            getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
        },
        __wbg_waitForTransaction_db322724fdb656c1: function(arg0, arg1) {
            const ret = arg0.waitForTransaction(arg1);
            return ret;
        },
        __wbg_wasmmanagedcoreclientreadonly_new: function(arg0) {
            const ret = WasmManagedCoreClientReadOnly.__wrap(arg0);
            return ret;
        },
        __wbindgen_cast_0000000000000001: function(arg0, arg1) {
            // Cast intrinsic for `Closure(Closure { dtor_idx: 1250, function: Function { arguments: [Externref], shim_idx: 1251, ret: Unit, inner_ret: Some(Unit) }, mutable: true }) -> Externref`.
            const ret = makeMutClosure(arg0, arg1, wasm.wasm_bindgen__closure__destroy__h5a276fa5db8d9a09, wasm_bindgen__convert__closures_____invoke__h5d6ec017d3e58121);
            return ret;
        },
        __wbindgen_cast_0000000000000002: function(arg0, arg1) {
            // Cast intrinsic for `Closure(Closure { dtor_idx: 729, function: Function { arguments: [], shim_idx: 730, ret: Unit, inner_ret: Some(Unit) }, mutable: true }) -> Externref`.
            const ret = makeMutClosure(arg0, arg1, wasm.wasm_bindgen__closure__destroy__h00ccb526a1f544c2, wasm_bindgen__convert__closures_____invoke__h4da206c1ff01fdcd);
            return ret;
        },
        __wbindgen_cast_0000000000000003: function(arg0) {
            // Cast intrinsic for `F64 -> Externref`.
            const ret = arg0;
            return ret;
        },
        __wbindgen_cast_0000000000000004: function(arg0) {
            // Cast intrinsic for `I64 -> Externref`.
            const ret = arg0;
            return ret;
        },
        __wbindgen_cast_0000000000000005: function(arg0, arg1) {
            // Cast intrinsic for `Ref(String) -> Externref`.
            const ret = getStringFromWasm0(arg0, arg1);
            return ret;
        },
        __wbindgen_cast_0000000000000006: function(arg0) {
            // Cast intrinsic for `U64 -> Externref`.
            const ret = BigInt.asUintN(64, arg0);
            return ret;
        },
        __wbindgen_cast_0000000000000007: function(arg0, arg1) {
            var v0 = getArrayU8FromWasm0(arg0, arg1).slice();
            wasm.__wbindgen_free(arg0, arg1 * 1, 1);
            // Cast intrinsic for `Vector(U8) -> Externref`.
            const ret = v0;
            return ret;
        },
        __wbindgen_init_externref_table: function() {
            const table = wasm.__wbindgen_externrefs;
            const offset = table.grow(4);
            table.set(0, undefined);
            table.set(offset + 0, undefined);
            table.set(offset + 1, null);
            table.set(offset + 2, true);
            table.set(offset + 3, false);
        },
    };
    return {
        __proto__: null,
        "./notarization_wasm_bg.js": import0,
    };
}

function wasm_bindgen__convert__closures_____invoke__h4da206c1ff01fdcd(arg0, arg1) {
    wasm.wasm_bindgen__convert__closures_____invoke__h4da206c1ff01fdcd(arg0, arg1);
}

function wasm_bindgen__convert__closures_____invoke__h5d6ec017d3e58121(arg0, arg1, arg2) {
    wasm.wasm_bindgen__convert__closures_____invoke__h5d6ec017d3e58121(arg0, arg1, arg2);
}

function wasm_bindgen__convert__closures_____invoke__h9ec99707e3310bed(arg0, arg1, arg2, arg3) {
    wasm.wasm_bindgen__convert__closures_____invoke__h9ec99707e3310bed(arg0, arg1, arg2, arg3);
}


const __wbindgen_enum_NotarizationMethod = ["Dynamic", "Locked"];


const __wbindgen_enum_RequestCache = ["default", "no-store", "reload", "no-cache", "force-cache", "only-if-cached"];


const __wbindgen_enum_RequestCredentials = ["omit", "same-origin", "include"];


const __wbindgen_enum_RequestMode = ["same-origin", "no-cors", "cors", "navigate"];


const __wbindgen_enum_TimeLockType = ["None", "UnlockAt", "UntilDestroyed"];
const CreateNotarizationDynamicFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_createnotarizationdynamic_free(ptr >>> 0, 1));
const CreateNotarizationLockedFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_createnotarizationlocked_free(ptr >>> 0, 1));
const DataFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_data_free(ptr >>> 0, 1));
const DefaultHttpClientFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_defaulthttpclient_free(ptr >>> 0, 1));
const DestroyNotarizationFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_destroynotarization_free(ptr >>> 0, 1));
const EmptyFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_empty_free(ptr >>> 0, 1));
const GasStationParamsFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_gasstationparams_free(ptr >>> 0, 1));
const ImmutableMetadataFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_immutablemetadata_free(ptr >>> 0, 1));
const LockMetadataFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_lockmetadata_free(ptr >>> 0, 1));
const NotarizationBuilderDynamicFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_notarizationbuilderdynamic_free(ptr >>> 0, 1));
const NotarizationBuilderLockedFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_notarizationbuilderlocked_free(ptr >>> 0, 1));
const NotarizationClientFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_notarizationclient_free(ptr >>> 0, 1));
const NotarizationClientReadOnlyFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_notarizationclientreadonly_free(ptr >>> 0, 1));
const NotarizationResourceBuilderFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_notarizationresourcebuilder_free(ptr >>> 0, 1));
const OnChainNotarizationFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_onchainnotarization_free(ptr >>> 0, 1));
const StateFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_state_free(ptr >>> 0, 1));
const TimeLockFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_timelock_free(ptr >>> 0, 1));
const TransactionBuilderFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_transactionbuilder_free(ptr >>> 0, 1));
const TransactionOutputFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_transactionoutput_free(ptr >>> 0, 1));
const TransferNotarizationFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_transfernotarization_free(ptr >>> 0, 1));
const UpdateMetadataFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_updatemetadata_free(ptr >>> 0, 1));
const UpdateStateFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_updatestate_free(ptr >>> 0, 1));
const WasmManagedCoreClientFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_wasmmanagedcoreclient_free(ptr >>> 0, 1));
const WasmManagedCoreClientReadOnlyFinalization = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(ptr => wasm.__wbg_wasmmanagedcoreclientreadonly_free(ptr >>> 0, 1));

function addToExternrefTable0(obj) {
    const idx = wasm.__externref_table_alloc();
    wasm.__wbindgen_externrefs.set(idx, obj);
    return idx;
}

function _assertClass(instance, klass) {
    if (!(instance instanceof klass)) {
        throw new Error(`expected instance of ${klass.name}`);
    }
}

const CLOSURE_DTORS = (typeof FinalizationRegistry === 'undefined')
    ? { register: () => {}, unregister: () => {} }
    : new FinalizationRegistry(state => state.dtor(state.a, state.b));

function debugString(val) {
    // primitive types
    const type = typeof val;
    if (type == 'number' || type == 'boolean' || val == null) {
        return  `${val}`;
    }
    if (type == 'string') {
        return `"${val}"`;
    }
    if (type == 'symbol') {
        const description = val.description;
        if (description == null) {
            return 'Symbol';
        } else {
            return `Symbol(${description})`;
        }
    }
    if (type == 'function') {
        const name = val.name;
        if (typeof name == 'string' && name.length > 0) {
            return `Function(${name})`;
        } else {
            return 'Function';
        }
    }
    // objects
    if (Array.isArray(val)) {
        const length = val.length;
        let debug = '[';
        if (length > 0) {
            debug += debugString(val[0]);
        }
        for(let i = 1; i < length; i++) {
            debug += ', ' + debugString(val[i]);
        }
        debug += ']';
        return debug;
    }
    // Test for built-in
    const builtInMatches = /\[object ([^\]]+)\]/.exec(toString.call(val));
    let className;
    if (builtInMatches && builtInMatches.length > 1) {
        className = builtInMatches[1];
    } else {
        // Failed to match the standard '[object ClassName]'
        return toString.call(val);
    }
    if (className == 'Object') {
        // we're a user defined class or Object
        // JSON.stringify avoids problems with cycles, and is generally much
        // easier than looping through ownProperties of `val`.
        try {
            return 'Object(' + JSON.stringify(val) + ')';
        } catch (_) {
            return 'Object';
        }
    }
    // errors
    if (val instanceof Error) {
        return `${val.name}: ${val.message}\n${val.stack}`;
    }
    // TODO we could test for more things here, like `Set`s and `Map`s.
    return className;
}

function getArrayJsValueFromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    const mem = getDataViewMemory0();
    const result = [];
    for (let i = ptr; i < ptr + 4 * len; i += 4) {
        result.push(wasm.__wbindgen_externrefs.get(mem.getUint32(i, true)));
    }
    wasm.__externref_drop_slice(ptr, len);
    return result;
}

function getArrayU8FromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return getUint8ArrayMemory0().subarray(ptr / 1, ptr / 1 + len);
}

let cachedDataViewMemory0 = null;
function getDataViewMemory0() {
    if (cachedDataViewMemory0 === null || cachedDataViewMemory0.buffer.detached === true || (cachedDataViewMemory0.buffer.detached === undefined && cachedDataViewMemory0.buffer !== wasm.memory.buffer)) {
        cachedDataViewMemory0 = new DataView(wasm.memory.buffer);
    }
    return cachedDataViewMemory0;
}

function getStringFromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return decodeText(ptr, len);
}

let cachedUint8ArrayMemory0 = null;
function getUint8ArrayMemory0() {
    if (cachedUint8ArrayMemory0 === null || cachedUint8ArrayMemory0.byteLength === 0) {
        cachedUint8ArrayMemory0 = new Uint8Array(wasm.memory.buffer);
    }
    return cachedUint8ArrayMemory0;
}

function handleError(f, args) {
    try {
        return f.apply(this, args);
    } catch (e) {
        const idx = addToExternrefTable0(e);
        wasm.__wbindgen_exn_store(idx);
    }
}

function isLikeNone(x) {
    return x === undefined || x === null;
}

function makeMutClosure(arg0, arg1, dtor, f) {
    const state = { a: arg0, b: arg1, cnt: 1, dtor };
    const real = (...args) => {

        // First up with a closure we increment the internal reference
        // count. This ensures that the Rust closure environment won't
        // be deallocated while we're invoking it.
        state.cnt++;
        const a = state.a;
        state.a = 0;
        try {
            return f(a, state.b, ...args);
        } finally {
            state.a = a;
            real._wbg_cb_unref();
        }
    };
    real._wbg_cb_unref = () => {
        if (--state.cnt === 0) {
            state.dtor(state.a, state.b);
            state.a = 0;
            CLOSURE_DTORS.unregister(state);
        }
    };
    CLOSURE_DTORS.register(real, state, state);
    return real;
}

function passArrayJsValueToWasm0(array, malloc) {
    const ptr = malloc(array.length * 4, 4) >>> 0;
    for (let i = 0; i < array.length; i++) {
        const add = addToExternrefTable0(array[i]);
        getDataViewMemory0().setUint32(ptr + 4 * i, add, true);
    }
    WASM_VECTOR_LEN = array.length;
    return ptr;
}

function passStringToWasm0(arg, malloc, realloc) {
    if (realloc === undefined) {
        const buf = cachedTextEncoder.encode(arg);
        const ptr = malloc(buf.length, 1) >>> 0;
        getUint8ArrayMemory0().subarray(ptr, ptr + buf.length).set(buf);
        WASM_VECTOR_LEN = buf.length;
        return ptr;
    }

    let len = arg.length;
    let ptr = malloc(len, 1) >>> 0;

    const mem = getUint8ArrayMemory0();

    let offset = 0;

    for (; offset < len; offset++) {
        const code = arg.charCodeAt(offset);
        if (code > 0x7F) break;
        mem[ptr + offset] = code;
    }
    if (offset !== len) {
        if (offset !== 0) {
            arg = arg.slice(offset);
        }
        ptr = realloc(ptr, len, len = offset + arg.length * 3, 1) >>> 0;
        const view = getUint8ArrayMemory0().subarray(ptr + offset, ptr + len);
        const ret = cachedTextEncoder.encodeInto(arg, view);

        offset += ret.written;
        ptr = realloc(ptr, len, offset, 1) >>> 0;
    }

    WASM_VECTOR_LEN = offset;
    return ptr;
}

function takeFromExternrefTable0(idx) {
    const value = wasm.__wbindgen_externrefs.get(idx);
    wasm.__externref_table_dealloc(idx);
    return value;
}

let cachedTextDecoder = new TextDecoder('utf-8', { ignoreBOM: true, fatal: true });
cachedTextDecoder.decode();
function decodeText(ptr, len) {
    return cachedTextDecoder.decode(getUint8ArrayMemory0().subarray(ptr, ptr + len));
}

const cachedTextEncoder = new TextEncoder();

if (!('encodeInto' in cachedTextEncoder)) {
    cachedTextEncoder.encodeInto = function (arg, view) {
        const buf = cachedTextEncoder.encode(arg);
        view.set(buf);
        return {
            read: arg.length,
            written: buf.length
        };
    };
}

let WASM_VECTOR_LEN = 0;

const wasmPath = `${__dirname}/notarization_wasm_bg.wasm`;
const wasmBytes = require('fs').readFileSync(wasmPath);
const wasmModule = new WebAssembly.Module(wasmBytes);
const wasm = new WebAssembly.Instance(wasmModule, __wbg_get_imports()).exports;
wasm.__wbindgen_start();
