// @iota/sdk-wasm/web/lib/bindings.js
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0
// Import needs to be in a single line, otherwise it breaks
// prettier-ignore
// @ts-ignore: path is set to match runtime transpiled js path when bundled.
import { initLogger, createClient, destroyClient, createSecretManager, createWallet, callClientMethodAsync, callSecretManagerMethodAsync, callUtilsMethodRust, callWalletMethodAsync, destroyWallet, listenWalletAsync, getClientFromWallet, getSecretManagerFromWallet, listenMqtt, migrateStrongholdSnapshotV2ToV3, migrateDbChrysalisToStardust } from '../wasm/iota_sdk_wasm';
var callUtilsMethod = function (method) {
    var response = JSON.parse(callUtilsMethodRust(JSON.stringify(method)));
    if (response.type == 'error' || response.type == 'panic') {
        throw response;
    }
    else {
        return response.payload;
    }
};
export { initLogger, createClient, createWallet, createSecretManager, callClientMethodAsync, callSecretManagerMethodAsync, callUtilsMethod, callWalletMethodAsync, listenWalletAsync, destroyWallet, destroyClient, getClientFromWallet, getSecretManagerFromWallet, listenMqtt, migrateStrongholdSnapshotV2ToV3, migrateDbChrysalisToStardust, };
